// feedsmith/dist/opml/parse/utils.cjs
const require_utils = require('../../common/utils.cjs');

//#region src/opml/parse/utils.ts
const parseOutline = (value, options) => {
	if (!require_utils.isObject(value)) return;
	const outline = {
		text: require_utils.parseString(value["@text"]),
		type: require_utils.parseString(value["@type"]),
		isComment: require_utils.parseBoolean(value["@iscomment"]),
		isBreakpoint: require_utils.parseBoolean(value["@isbreakpoint"]),
		created: require_utils.parseDate(value["@created"]),
		category: require_utils.parseString(value["@category"]),
		description: require_utils.parseString(value["@description"]),
		xmlUrl: require_utils.parseString(value["@xmlurl"]),
		htmlUrl: require_utils.parseString(value["@htmlurl"]),
		language: require_utils.parseString(value["@language"]),
		title: require_utils.parseString(value["@title"]),
		version: require_utils.parseString(value["@version"]),
		url: require_utils.parseString(value["@url"]),
		outlines: require_utils.parseArrayOf(value.outline, (value$1) => parseOutline(value$1, options))
	};
	if (options?.extraOutlineAttributes) for (const attribute of options.extraOutlineAttributes) {
		const attributeKey = `@${attribute.toLowerCase()}`;
		if (!require_utils.isPresent(value[attributeKey])) continue;
		outline[attribute] = require_utils.parseString(value[attributeKey]);
	}
	return require_utils.trimObject(outline);
};
const parseHead = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		title: require_utils.parseSingularOf(value.title, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		dateCreated: require_utils.parseSingularOf(value.datecreated, (value$1) => require_utils.parseDate(require_utils.retrieveText(value$1))),
		dateModified: require_utils.parseSingularOf(value.datemodified, (value$1) => require_utils.parseDate(require_utils.retrieveText(value$1))),
		ownerName: require_utils.parseSingularOf(value.ownername, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		ownerEmail: require_utils.parseSingularOf(value.owneremail, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		ownerId: require_utils.parseSingularOf(value.ownerid, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		docs: require_utils.parseSingularOf(value.docs, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		expansionState: require_utils.parseSingularOf(value.expansionstate, (value$1) => require_utils.parseCsvOf(require_utils.retrieveText(value$1), require_utils.parseNumber)),
		vertScrollState: require_utils.parseSingularOf(value.vertscrollstate, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		windowTop: require_utils.parseSingularOf(value.windowtop, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		windowLeft: require_utils.parseSingularOf(value.windowleft, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		windowBottom: require_utils.parseSingularOf(value.windowbottom, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		windowRight: require_utils.parseSingularOf(value.windowright, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1)))
	});
};
const parseBody = (value, options) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({ outlines: require_utils.parseArrayOf(value.outline, (value$1) => parseOutline(value$1, options), options?.maxItems) });
};
const parseDocument = (value, options) => {
	if (!require_utils.isObject(value?.opml)) return;
	return require_utils.trimObject({
		head: parseHead(value.opml.head),
		body: parseBody(value.opml.body, options)
	});
};

//#endregion
exports.parseDocument = parseDocument;
