// feedsmith/dist/opml/parse/utils.js
import { isObject, isPresent, parseArrayOf, parseBoolean, parseCsvOf, parseDate, parseNumber, parseSingularOf, parseString, retrieveText, trimObject } from "../../common/utils.js";

//#region src/opml/parse/utils.ts
const parseOutline = (value, options) => {
	if (!isObject(value)) return;
	const outline = {
		text: parseString(value["@text"]),
		type: parseString(value["@type"]),
		isComment: parseBoolean(value["@iscomment"]),
		isBreakpoint: parseBoolean(value["@isbreakpoint"]),
		created: parseDate(value["@created"]),
		category: parseString(value["@category"]),
		description: parseString(value["@description"]),
		xmlUrl: parseString(value["@xmlurl"]),
		htmlUrl: parseString(value["@htmlurl"]),
		language: parseString(value["@language"]),
		title: parseString(value["@title"]),
		version: parseString(value["@version"]),
		url: parseString(value["@url"]),
		outlines: parseArrayOf(value.outline, (value$1) => parseOutline(value$1, options))
	};
	if (options?.extraOutlineAttributes) for (const attribute of options.extraOutlineAttributes) {
		const attributeKey = `@${attribute.toLowerCase()}`;
		if (!isPresent(value[attributeKey])) continue;
		outline[attribute] = parseString(value[attributeKey]);
	}
	return trimObject(outline);
};
const parseHead = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		title: parseSingularOf(value.title, (value$1) => parseString(retrieveText(value$1))),
		dateCreated: parseSingularOf(value.datecreated, (value$1) => parseDate(retrieveText(value$1))),
		dateModified: parseSingularOf(value.datemodified, (value$1) => parseDate(retrieveText(value$1))),
		ownerName: parseSingularOf(value.ownername, (value$1) => parseString(retrieveText(value$1))),
		ownerEmail: parseSingularOf(value.owneremail, (value$1) => parseString(retrieveText(value$1))),
		ownerId: parseSingularOf(value.ownerid, (value$1) => parseString(retrieveText(value$1))),
		docs: parseSingularOf(value.docs, (value$1) => parseString(retrieveText(value$1))),
		expansionState: parseSingularOf(value.expansionstate, (value$1) => parseCsvOf(retrieveText(value$1), parseNumber)),
		vertScrollState: parseSingularOf(value.vertscrollstate, (value$1) => parseNumber(retrieveText(value$1))),
		windowTop: parseSingularOf(value.windowtop, (value$1) => parseNumber(retrieveText(value$1))),
		windowLeft: parseSingularOf(value.windowleft, (value$1) => parseNumber(retrieveText(value$1))),
		windowBottom: parseSingularOf(value.windowbottom, (value$1) => parseNumber(retrieveText(value$1))),
		windowRight: parseSingularOf(value.windowright, (value$1) => parseNumber(retrieveText(value$1)))
	});
};
const parseBody = (value, options) => {
	if (!isObject(value)) return;
	return trimObject({ outlines: parseArrayOf(value.outline, (value$1) => parseOutline(value$1, options), options?.maxItems) });
};
const parseDocument = (value, options) => {
	if (!isObject(value?.opml)) return;
	return trimObject({
		head: parseHead(value.opml.head),
		body: parseBody(value.opml.body, options)
	});
};

//#endregion
export { parseDocument };
