// feedsmith/dist/namespaces/arxiv/parse/utils.js
import { isObject, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/arxiv/parse/utils.ts
const parsePrimaryCategory = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		term: parseString(value["@term"]),
		scheme: parseString(value["@scheme"]),
		label: parseString(value["@label"])
	});
};
const retrieveAuthor = (value) => {
	if (!isObject(value)) return;
	return trimObject({ affiliation: parseSingularOf(value["arxiv:affiliation"], (value$1) => parseString(retrieveText(value$1))) });
};
const retrieveEntry = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		comment: parseSingularOf(value["arxiv:comment"], (value$1) => parseString(retrieveText(value$1))),
		journalRef: parseSingularOf(value["arxiv:journal_ref"], (value$1) => parseString(retrieveText(value$1))),
		doi: parseSingularOf(value["arxiv:doi"], (value$1) => parseString(retrieveText(value$1))),
		primaryCategory: parseSingularOf(value["arxiv:primary_category"], parsePrimaryCategory)
	});
};

//#endregion
export { retrieveAuthor, retrieveEntry };
