// feedsmith/dist/namespaces/arxiv/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/arxiv/parse/utils.ts
const parsePrimaryCategory = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		term: require_utils.parseString(value["@term"]),
		scheme: require_utils.parseString(value["@scheme"]),
		label: require_utils.parseString(value["@label"])
	});
};
const retrieveAuthor = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({ affiliation: require_utils.parseSingularOf(value["arxiv:affiliation"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))) });
};
const retrieveEntry = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		comment: require_utils.parseSingularOf(value["arxiv:comment"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		journalRef: require_utils.parseSingularOf(value["arxiv:journal_ref"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		doi: require_utils.parseSingularOf(value["arxiv:doi"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		primaryCategory: require_utils.parseSingularOf(value["arxiv:primary_category"], parsePrimaryCategory)
	});
};

//#endregion
exports.retrieveAuthor = retrieveAuthor;
exports.retrieveEntry = retrieveEntry;
