// feedsmith/dist/feeds/rss/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');
const require_utils$1 = require('../../../namespaces/admin/parse/utils.cjs');
const require_utils$2 = require('../../../namespaces/cc/parse/utils.cjs');
const require_utils$3 = require('../../../namespaces/creativecommons/parse/utils.cjs');
const require_utils$4 = require('../../../namespaces/dc/parse/utils.cjs');
const require_utils$5 = require('../../../namespaces/dcterms/parse/utils.cjs');
const require_utils$6 = require('../../../namespaces/geo/parse/utils.cjs');
const require_utils$7 = require('../../../namespaces/georss/parse/utils.cjs');
const require_utils$8 = require('../../../namespaces/googleplay/parse/utils.cjs');
const require_utils$9 = require('../../../namespaces/itunes/parse/utils.cjs');
const require_utils$10 = require('../../../namespaces/media/parse/utils.cjs');
const require_utils$11 = require('../../../namespaces/opensearch/parse/utils.cjs');
const require_utils$12 = require('../../../namespaces/pingback/parse/utils.cjs');
const require_utils$13 = require('../../../namespaces/psc/parse/utils.cjs');
const require_utils$14 = require('../../../namespaces/slash/parse/utils.cjs');
const require_utils$15 = require('../../../namespaces/sy/parse/utils.cjs');
const require_utils$16 = require('../../../namespaces/thr/parse/utils.cjs');
const require_utils$17 = require('../../../namespaces/trackback/parse/utils.cjs');
const require_utils$18 = require('../../../namespaces/wfw/parse/utils.cjs');
const require_utils$19 = require('../../../namespaces/atom/parse/utils.cjs');
const require_utils$20 = require('../../../namespaces/content/parse/utils.cjs');
const require_utils$21 = require('../../../namespaces/acast/parse/utils.cjs');
const require_utils$22 = require('../../../namespaces/blogchannel/parse/utils.cjs');
const require_utils$23 = require('../../../namespaces/feedpress/parse/utils.cjs');
const require_utils$24 = require('../../../namespaces/podcast/parse/utils.cjs');
const require_utils$25 = require('../../../namespaces/prism/parse/utils.cjs');
const require_utils$26 = require('../../../namespaces/rawvoice/parse/utils.cjs');
const require_utils$27 = require('../../../namespaces/source/parse/utils.cjs');
const require_utils$28 = require('../../../namespaces/spotify/parse/utils.cjs');

//#region src/feeds/rss/parse/utils.ts
const parsePerson = (value) => {
	return require_utils.parseSingularOf(value?.name ?? value, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)));
};
const parseCategory = (value) => {
	return require_utils.trimObject({
		name: require_utils.parseString(require_utils.retrieveText(value)),
		domain: require_utils.parseString(value?.["@domain"])
	});
};
const parseCloud = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		domain: require_utils.parseString(value["@domain"]),
		port: require_utils.parseNumber(value["@port"]),
		path: require_utils.parseString(value["@path"]),
		registerProcedure: require_utils.parseString(value["@registerprocedure"]),
		protocol: require_utils.parseString(value["@protocol"])
	});
};
const parseImage = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		url: require_utils.parseSingularOf(value.url, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		title: require_utils.parseSingularOf(value.title, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		link: require_utils.parseSingularOf(value.link, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		description: require_utils.parseSingularOf(value.description, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		height: require_utils.parseSingularOf(value.height, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		width: require_utils.parseSingularOf(value.width, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1)))
	});
};
const parseTextInput = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		title: require_utils.parseSingularOf(value.title, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		description: require_utils.parseSingularOf(value.description, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		name: require_utils.parseSingularOf(value.name, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		link: require_utils.parseSingularOf(value.link, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};
const parseSkipHours = (value) => {
	return require_utils.trimArray(value?.hour, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1)));
};
const parseSkipDays = (value) => {
	return require_utils.trimArray(value?.day, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)));
};
const parseEnclosure = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		url: require_utils.parseString(value["@url"]),
		length: require_utils.parseNumber(value["@length"]),
		type: require_utils.parseString(value["@type"])
	});
};
const parseGuid = (value) => {
	return require_utils.trimObject({
		value: require_utils.parseString(require_utils.retrieveText(value)),
		isPermaLink: require_utils.parseBoolean(value?.["@ispermalink"])
	});
};
const parseSource = (value) => {
	return require_utils.trimObject({
		title: require_utils.parseString(require_utils.retrieveText(value)),
		url: require_utils.parseString(value?.["@url"])
	});
};
const parseItem = (value) => {
	if (!require_utils.isObject(value)) return;
	const namespaces = require_utils.detectNamespaces(value);
	return require_utils.trimObject({
		title: require_utils.parseSingularOf(value.title, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		link: require_utils.parseSingularOf(value.link, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		description: require_utils.parseSingularOf(value.description, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		authors: require_utils.parseArrayOf(value.author, parsePerson),
		categories: require_utils.parseArrayOf(value.category, parseCategory),
		comments: require_utils.parseSingularOf(value.comments, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		enclosures: require_utils.parseArrayOf(value.enclosure, parseEnclosure),
		guid: require_utils.parseSingularOf(value.guid, parseGuid),
		pubDate: require_utils.parseSingularOf(value.pubdate, (value$1) => require_utils.parseDate(require_utils.retrieveText(value$1))),
		source: require_utils.parseSingularOf(value.source, parseSource),
		atom: namespaces.has("atom") ? require_utils$19.retrieveEntry(value) : void 0,
		cc: namespaces.has("cc") ? require_utils$2.retrieveItemOrFeed(value) : void 0,
		dc: namespaces.has("dc") ? require_utils$4.retrieveItemOrFeed(value) : void 0,
		content: namespaces.has("content") ? require_utils$20.retrieveItem(value) : void 0,
		creativeCommons: namespaces.has("creativecommons") ? require_utils$3.retrieveItemOrFeed(value) : void 0,
		slash: namespaces.has("slash") ? require_utils$14.retrieveItem(value) : void 0,
		itunes: namespaces.has("itunes") ? require_utils$9.retrieveItem(value) : void 0,
		podcast: namespaces.has("podcast") ? require_utils$24.retrieveItem(value) : void 0,
		psc: namespaces.has("psc") ? require_utils$13.retrieveItem(value) : void 0,
		googleplay: namespaces.has("googleplay") ? require_utils$8.retrieveItem(value) : void 0,
		media: namespaces.has("media") ? require_utils$10.retrieveItemOrFeed(value) : void 0,
		georss: namespaces.has("georss") ? require_utils$7.retrieveItemOrFeed(value) : void 0,
		geo: namespaces.has("geo") ? require_utils$6.retrieveItemOrFeed(value) : void 0,
		thr: namespaces.has("thr") ? require_utils$16.retrieveItem(value) : void 0,
		dcterms: namespaces.has("dcterms") ? require_utils$5.retrieveItemOrFeed(value) : void 0,
		prism: namespaces.has("prism") ? require_utils$25.retrieveItem(value) : void 0,
		wfw: namespaces.has("wfw") ? require_utils$18.retrieveItem(value) : void 0,
		sourceNs: namespaces.has("source") ? require_utils$27.retrieveItem(value) : void 0,
		rawvoice: namespaces.has("rawvoice") ? require_utils$26.retrieveItem(value) : void 0,
		spotify: namespaces.has("spotify") ? require_utils$28.retrieveItem(value) : void 0,
		pingback: namespaces.has("pingback") ? require_utils$12.retrieveItem(value) : void 0,
		trackback: namespaces.has("trackback") ? require_utils$17.retrieveItem(value) : void 0,
		acast: namespaces.has("acast") ? require_utils$21.retrieveItem(value) : void 0
	});
};
const parseFeed = (value, options) => {
	if (!require_utils.isObject(value)) return;
	const namespaces = require_utils.detectNamespaces(value);
	return require_utils.trimObject({
		title: require_utils.parseSingularOf(value.title, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		link: require_utils.parseSingularOf(value.link, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		description: require_utils.parseSingularOf(value.description, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		language: require_utils.parseSingularOf(value.language, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		copyright: require_utils.parseSingularOf(value.copyright, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		managingEditor: require_utils.parseSingularOf(value.managingeditor, parsePerson),
		webMaster: require_utils.parseSingularOf(value.webmaster, parsePerson),
		pubDate: require_utils.parseSingularOf(value.pubdate, (value$1) => require_utils.parseDate(require_utils.retrieveText(value$1))),
		lastBuildDate: require_utils.parseSingularOf(value.lastbuilddate, (value$1) => require_utils.parseDate(require_utils.retrieveText(value$1))),
		categories: require_utils.parseArrayOf(value.category, parseCategory),
		generator: require_utils.parseSingularOf(value.generator, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		docs: require_utils.parseSingularOf(value.docs, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		cloud: require_utils.parseSingularOf(value.cloud, parseCloud),
		ttl: require_utils.parseSingularOf(value.ttl, (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		image: require_utils.parseSingularOf(value.image, parseImage),
		rating: require_utils.parseSingularOf(value.rating, (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		textInput: require_utils.parseSingularOf(value.textinput, parseTextInput),
		skipHours: require_utils.parseSingularOf(value.skiphours, parseSkipHours),
		skipDays: require_utils.parseSingularOf(value.skipdays, parseSkipDays),
		items: require_utils.parseArrayOf(value.item, parseItem, options?.maxItems),
		atom: namespaces.has("atom") ? require_utils$19.retrieveFeed(value) : void 0,
		cc: namespaces.has("cc") ? require_utils$2.retrieveItemOrFeed(value) : void 0,
		dc: namespaces.has("dc") ? require_utils$4.retrieveItemOrFeed(value) : void 0,
		sy: namespaces.has("sy") ? require_utils$15.retrieveFeed(value) : void 0,
		itunes: namespaces.has("itunes") ? require_utils$9.retrieveFeed(value) : void 0,
		podcast: namespaces.has("podcast") ? require_utils$24.retrieveFeed(value) : void 0,
		googleplay: namespaces.has("googleplay") ? require_utils$8.retrieveFeed(value) : void 0,
		media: namespaces.has("media") ? require_utils$10.retrieveItemOrFeed(value) : void 0,
		georss: namespaces.has("georss") ? require_utils$7.retrieveItemOrFeed(value) : void 0,
		geo: namespaces.has("geo") ? require_utils$6.retrieveItemOrFeed(value) : void 0,
		dcterms: namespaces.has("dcterms") ? require_utils$5.retrieveItemOrFeed(value) : void 0,
		prism: namespaces.has("prism") ? require_utils$25.retrieveFeed(value) : void 0,
		creativeCommons: namespaces.has("creativecommons") ? require_utils$3.retrieveItemOrFeed(value) : void 0,
		feedpress: namespaces.has("feedpress") ? require_utils$23.retrieveFeed(value) : void 0,
		opensearch: namespaces.has("opensearch") ? require_utils$11.retrieveFeed(value) : void 0,
		admin: namespaces.has("admin") ? require_utils$1.retrieveFeed(value) : void 0,
		sourceNs: namespaces.has("source") ? require_utils$27.retrieveFeed(value) : void 0,
		blogChannel: namespaces.has("blogchannel") ? require_utils$22.retrieveFeed(value) : void 0,
		rawvoice: namespaces.has("rawvoice") ? require_utils$26.retrieveFeed(value) : void 0,
		spotify: namespaces.has("spotify") ? require_utils$28.retrieveFeed(value) : void 0,
		pingback: namespaces.has("pingback") ? require_utils$12.retrieveFeed(value) : void 0,
		acast: namespaces.has("acast") ? require_utils$21.retrieveFeed(value) : void 0
	});
};
const retrieveFeed = (value, options) => {
	return require_utils.parseSingularOf(value?.rss?.channel, (value$1) => parseFeed(value$1, options));
};

//#endregion
exports.retrieveFeed = retrieveFeed;
