// feedsmith/dist/feeds/rss/parse/utils.js
import { detectNamespaces, isObject, parseArrayOf, parseBoolean, parseDate, parseNumber, parseSingularOf, parseString, retrieveText, trimArray, trimObject } from "../../../common/utils.js";
import { retrieveFeed as retrieveFeed$1 } from "../../../namespaces/admin/parse/utils.js";
import { retrieveItemOrFeed } from "../../../namespaces/cc/parse/utils.js";
import { retrieveItemOrFeed as retrieveItemOrFeed$1 } from "../../../namespaces/creativecommons/parse/utils.js";
import { retrieveItemOrFeed as retrieveItemOrFeed$2 } from "../../../namespaces/dc/parse/utils.js";
import { retrieveItemOrFeed as retrieveItemOrFeed$3 } from "../../../namespaces/dcterms/parse/utils.js";
import { retrieveItemOrFeed as retrieveItemOrFeed$4 } from "../../../namespaces/geo/parse/utils.js";
import { retrieveItemOrFeed as retrieveItemOrFeed$5 } from "../../../namespaces/georss/parse/utils.js";
import { retrieveFeed as retrieveFeed$2, retrieveItem } from "../../../namespaces/googleplay/parse/utils.js";
import { retrieveFeed as retrieveFeed$3, retrieveItem as retrieveItem$1 } from "../../../namespaces/itunes/parse/utils.js";
import { retrieveItemOrFeed as retrieveItemOrFeed$6 } from "../../../namespaces/media/parse/utils.js";
import { retrieveFeed as retrieveFeed$4 } from "../../../namespaces/opensearch/parse/utils.js";
import { retrieveFeed as retrieveFeed$5, retrieveItem as retrieveItem$2 } from "../../../namespaces/pingback/parse/utils.js";
import { retrieveItem as retrieveItem$3 } from "../../../namespaces/psc/parse/utils.js";
import { retrieveItem as retrieveItem$4 } from "../../../namespaces/slash/parse/utils.js";
import { retrieveFeed as retrieveFeed$6 } from "../../../namespaces/sy/parse/utils.js";
import { retrieveItem as retrieveItem$5 } from "../../../namespaces/thr/parse/utils.js";
import { retrieveItem as retrieveItem$6 } from "../../../namespaces/trackback/parse/utils.js";
import { retrieveItem as retrieveItem$7 } from "../../../namespaces/wfw/parse/utils.js";
import { retrieveEntry, retrieveFeed as retrieveFeed$7 } from "../../../namespaces/atom/parse/utils.js";
import { retrieveItem as retrieveItem$8 } from "../../../namespaces/content/parse/utils.js";
import { retrieveFeed as retrieveFeed$8, retrieveItem as retrieveItem$9 } from "../../../namespaces/acast/parse/utils.js";
import { retrieveFeed as retrieveFeed$9 } from "../../../namespaces/blogchannel/parse/utils.js";
import { retrieveFeed as retrieveFeed$10 } from "../../../namespaces/feedpress/parse/utils.js";
import { retrieveFeed as retrieveFeed$11, retrieveItem as retrieveItem$10 } from "../../../namespaces/podcast/parse/utils.js";
import { retrieveFeed as retrieveFeed$12, retrieveItem as retrieveItem$11 } from "../../../namespaces/prism/parse/utils.js";
import { retrieveFeed as retrieveFeed$13, retrieveItem as retrieveItem$12 } from "../../../namespaces/rawvoice/parse/utils.js";
import { retrieveFeed as retrieveFeed$14, retrieveItem as retrieveItem$13 } from "../../../namespaces/source/parse/utils.js";
import { retrieveFeed as retrieveFeed$15, retrieveItem as retrieveItem$14 } from "../../../namespaces/spotify/parse/utils.js";

//#region src/feeds/rss/parse/utils.ts
const parsePerson = (value) => {
	return parseSingularOf(value?.name ?? value, (value$1) => parseString(retrieveText(value$1)));
};
const parseCategory = (value) => {
	return trimObject({
		name: parseString(retrieveText(value)),
		domain: parseString(value?.["@domain"])
	});
};
const parseCloud = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		domain: parseString(value["@domain"]),
		port: parseNumber(value["@port"]),
		path: parseString(value["@path"]),
		registerProcedure: parseString(value["@registerprocedure"]),
		protocol: parseString(value["@protocol"])
	});
};
const parseImage = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		url: parseSingularOf(value.url, (value$1) => parseString(retrieveText(value$1))),
		title: parseSingularOf(value.title, (value$1) => parseString(retrieveText(value$1))),
		link: parseSingularOf(value.link, (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value.description, (value$1) => parseString(retrieveText(value$1))),
		height: parseSingularOf(value.height, (value$1) => parseNumber(retrieveText(value$1))),
		width: parseSingularOf(value.width, (value$1) => parseNumber(retrieveText(value$1)))
	});
};
const parseTextInput = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		title: parseSingularOf(value.title, (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value.description, (value$1) => parseString(retrieveText(value$1))),
		name: parseSingularOf(value.name, (value$1) => parseString(retrieveText(value$1))),
		link: parseSingularOf(value.link, (value$1) => parseString(retrieveText(value$1)))
	});
};
const parseSkipHours = (value) => {
	return trimArray(value?.hour, (value$1) => parseNumber(retrieveText(value$1)));
};
const parseSkipDays = (value) => {
	return trimArray(value?.day, (value$1) => parseString(retrieveText(value$1)));
};
const parseEnclosure = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		url: parseString(value["@url"]),
		length: parseNumber(value["@length"]),
		type: parseString(value["@type"])
	});
};
const parseGuid = (value) => {
	return trimObject({
		value: parseString(retrieveText(value)),
		isPermaLink: parseBoolean(value?.["@ispermalink"])
	});
};
const parseSource = (value) => {
	return trimObject({
		title: parseString(retrieveText(value)),
		url: parseString(value?.["@url"])
	});
};
const parseItem = (value) => {
	if (!isObject(value)) return;
	const namespaces = detectNamespaces(value);
	return trimObject({
		title: parseSingularOf(value.title, (value$1) => parseString(retrieveText(value$1))),
		link: parseSingularOf(value.link, (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value.description, (value$1) => parseString(retrieveText(value$1))),
		authors: parseArrayOf(value.author, parsePerson),
		categories: parseArrayOf(value.category, parseCategory),
		comments: parseSingularOf(value.comments, (value$1) => parseString(retrieveText(value$1))),
		enclosures: parseArrayOf(value.enclosure, parseEnclosure),
		guid: parseSingularOf(value.guid, parseGuid),
		pubDate: parseSingularOf(value.pubdate, (value$1) => parseDate(retrieveText(value$1))),
		source: parseSingularOf(value.source, parseSource),
		atom: namespaces.has("atom") ? retrieveEntry(value) : void 0,
		cc: namespaces.has("cc") ? retrieveItemOrFeed(value) : void 0,
		dc: namespaces.has("dc") ? retrieveItemOrFeed$2(value) : void 0,
		content: namespaces.has("content") ? retrieveItem$8(value) : void 0,
		creativeCommons: namespaces.has("creativecommons") ? retrieveItemOrFeed$1(value) : void 0,
		slash: namespaces.has("slash") ? retrieveItem$4(value) : void 0,
		itunes: namespaces.has("itunes") ? retrieveItem$1(value) : void 0,
		podcast: namespaces.has("podcast") ? retrieveItem$10(value) : void 0,
		psc: namespaces.has("psc") ? retrieveItem$3(value) : void 0,
		googleplay: namespaces.has("googleplay") ? retrieveItem(value) : void 0,
		media: namespaces.has("media") ? retrieveItemOrFeed$6(value) : void 0,
		georss: namespaces.has("georss") ? retrieveItemOrFeed$5(value) : void 0,
		geo: namespaces.has("geo") ? retrieveItemOrFeed$4(value) : void 0,
		thr: namespaces.has("thr") ? retrieveItem$5(value) : void 0,
		dcterms: namespaces.has("dcterms") ? retrieveItemOrFeed$3(value) : void 0,
		prism: namespaces.has("prism") ? retrieveItem$11(value) : void 0,
		wfw: namespaces.has("wfw") ? retrieveItem$7(value) : void 0,
		sourceNs: namespaces.has("source") ? retrieveItem$13(value) : void 0,
		rawvoice: namespaces.has("rawvoice") ? retrieveItem$12(value) : void 0,
		spotify: namespaces.has("spotify") ? retrieveItem$14(value) : void 0,
		pingback: namespaces.has("pingback") ? retrieveItem$2(value) : void 0,
		trackback: namespaces.has("trackback") ? retrieveItem$6(value) : void 0,
		acast: namespaces.has("acast") ? retrieveItem$9(value) : void 0
	});
};
const parseFeed = (value, options) => {
	if (!isObject(value)) return;
	const namespaces = detectNamespaces(value);
	return trimObject({
		title: parseSingularOf(value.title, (value$1) => parseString(retrieveText(value$1))),
		link: parseSingularOf(value.link, (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value.description, (value$1) => parseString(retrieveText(value$1))),
		language: parseSingularOf(value.language, (value$1) => parseString(retrieveText(value$1))),
		copyright: parseSingularOf(value.copyright, (value$1) => parseString(retrieveText(value$1))),
		managingEditor: parseSingularOf(value.managingeditor, parsePerson),
		webMaster: parseSingularOf(value.webmaster, parsePerson),
		pubDate: parseSingularOf(value.pubdate, (value$1) => parseDate(retrieveText(value$1))),
		lastBuildDate: parseSingularOf(value.lastbuilddate, (value$1) => parseDate(retrieveText(value$1))),
		categories: parseArrayOf(value.category, parseCategory),
		generator: parseSingularOf(value.generator, (value$1) => parseString(retrieveText(value$1))),
		docs: parseSingularOf(value.docs, (value$1) => parseString(retrieveText(value$1))),
		cloud: parseSingularOf(value.cloud, parseCloud),
		ttl: parseSingularOf(value.ttl, (value$1) => parseNumber(retrieveText(value$1))),
		image: parseSingularOf(value.image, parseImage),
		rating: parseSingularOf(value.rating, (value$1) => parseString(retrieveText(value$1))),
		textInput: parseSingularOf(value.textinput, parseTextInput),
		skipHours: parseSingularOf(value.skiphours, parseSkipHours),
		skipDays: parseSingularOf(value.skipdays, parseSkipDays),
		items: parseArrayOf(value.item, parseItem, options?.maxItems),
		atom: namespaces.has("atom") ? retrieveFeed$7(value) : void 0,
		cc: namespaces.has("cc") ? retrieveItemOrFeed(value) : void 0,
		dc: namespaces.has("dc") ? retrieveItemOrFeed$2(value) : void 0,
		sy: namespaces.has("sy") ? retrieveFeed$6(value) : void 0,
		itunes: namespaces.has("itunes") ? retrieveFeed$3(value) : void 0,
		podcast: namespaces.has("podcast") ? retrieveFeed$11(value) : void 0,
		googleplay: namespaces.has("googleplay") ? retrieveFeed$2(value) : void 0,
		media: namespaces.has("media") ? retrieveItemOrFeed$6(value) : void 0,
		georss: namespaces.has("georss") ? retrieveItemOrFeed$5(value) : void 0,
		geo: namespaces.has("geo") ? retrieveItemOrFeed$4(value) : void 0,
		dcterms: namespaces.has("dcterms") ? retrieveItemOrFeed$3(value) : void 0,
		prism: namespaces.has("prism") ? retrieveFeed$12(value) : void 0,
		creativeCommons: namespaces.has("creativecommons") ? retrieveItemOrFeed$1(value) : void 0,
		feedpress: namespaces.has("feedpress") ? retrieveFeed$10(value) : void 0,
		opensearch: namespaces.has("opensearch") ? retrieveFeed$4(value) : void 0,
		admin: namespaces.has("admin") ? retrieveFeed$1(value) : void 0,
		sourceNs: namespaces.has("source") ? retrieveFeed$14(value) : void 0,
		blogChannel: namespaces.has("blogchannel") ? retrieveFeed$9(value) : void 0,
		rawvoice: namespaces.has("rawvoice") ? retrieveFeed$13(value) : void 0,
		spotify: namespaces.has("spotify") ? retrieveFeed$15(value) : void 0,
		pingback: namespaces.has("pingback") ? retrieveFeed$5(value) : void 0,
		acast: namespaces.has("acast") ? retrieveFeed$8(value) : void 0
	});
};
const retrieveFeed = (value, options) => {
	return parseSingularOf(value?.rss?.channel, (value$1) => parseFeed(value$1, options));
};

//#endregion
export { retrieveFeed };
