// feedsmith/dist/common/config.cjs

//#region src/common/config.ts
const parserConfig = {
	trimValues: false,
	processEntities: false,
	htmlEntities: false,
	parseTagValue: false,
	parseAttributeValue: false,
	alwaysCreateTextNode: false,
	ignoreAttributes: false,
	ignorePiTags: true,
	ignoreDeclaration: true,
	attributeNamePrefix: "@",
	transformTagName: (name) => name.toLowerCase(),
	transformAttributeName: (name) => name.toLowerCase()
};
const builderConfig = {
	processEntities: true,
	ignoreAttributes: false,
	suppressEmptyNode: true,
	suppressBooleanAttributes: false,
	attributeNamePrefix: "@",
	format: true,
	cdataPropName: "#cdata"
};
const locales = {
	unrecognizedFeedFormat: "Unrecognized feed format",
	invalidFeedFormat: "Invalid feed format",
	invalidOpmlFormat: "Invalid OPML format",
	invalidInputOpml: "Invalid input OPML",
	invalidInputAtom: "Invalid input Atom",
	invalidInputRss: "Invalid input RSS"
};
const namespaceUris = {
	admin: [
		"http://webns.net/mvcb/",
		"https://webns.net/mvcb/",
		"http://webns.net/mvcb",
		"https://webns.net/mvcb"
	],
	atom: [
		"http://www.w3.org/2005/Atom",
		"https://www.w3.org/2005/Atom",
		"http://www.w3.org/2005/Atom/",
		"https://www.w3.org/2005/Atom/",
		"http://purl.org/atom/ns#",
		"https://purl.org/atom/ns#"
	],
	blogChannel: [
		"http://backend.userland.com/blogChannelModule",
		"https://backend.userland.com/blogChannelModule",
		"http://backend.userland.com/blogChannelModule/",
		"https://backend.userland.com/blogChannelModule/"
	],
	app: [
		"http://www.w3.org/2007/app",
		"https://www.w3.org/2007/app",
		"http://www.w3.org/2007/app/",
		"https://www.w3.org/2007/app/"
	],
	dc: [
		"http://purl.org/dc/elements/1.1/",
		"https://purl.org/dc/elements/1.1/",
		"http://purl.org/dc/elements/1.1",
		"https://purl.org/dc/elements/1.1",
		"http://dublincore.org/documents/dcmi-namespace/",
		"https://dublincore.org/documents/dcmi-namespace/",
		"http://dublincore.org/documents/dcmi-namespace",
		"https://dublincore.org/documents/dcmi-namespace",
		"http://purl.org/dc/elements/1.0/",
		"https://purl.org/dc/elements/1.0/",
		"http://purl.org/dc/elements/1.0",
		"https://purl.org/dc/elements/1.0"
	],
	sy: [
		"http://purl.org/rss/1.0/modules/syndication/",
		"https://purl.org/rss/1.0/modules/syndication/",
		"http://purl.org/rss/1.0/modules/syndication",
		"https://purl.org/rss/1.0/modules/syndication"
	],
	content: [
		"http://purl.org/rss/1.0/modules/content/",
		"https://purl.org/rss/1.0/modules/content/",
		"http://purl.org/rss/1.0/modules/content",
		"https://purl.org/rss/1.0/modules/content"
	],
	creativeCommons: [
		"http://backend.userland.com/creativeCommonsRssModule",
		"https://backend.userland.com/creativeCommonsRssModule",
		"http://backend.userland.com/creativeCommonsRssModule/",
		"https://backend.userland.com/creativeCommonsRssModule/",
		"http://cyber.law.harvard.edu/rss/creativeCommonsRssModule.html",
		"https://cyber.law.harvard.edu/rss/creativeCommonsRssModule.html",
		"http://cyber.law.harvard.edu/rss/creativeCommonsRssModule",
		"https://cyber.law.harvard.edu/rss/creativeCommonsRssModule"
	],
	slash: [
		"http://purl.org/rss/1.0/modules/slash/",
		"https://purl.org/rss/1.0/modules/slash/",
		"http://purl.org/rss/1.0/modules/slash",
		"https://purl.org/rss/1.0/modules/slash"
	],
	itunes: ["http://www.itunes.com/dtds/podcast-1.0.dtd", "https://www.itunes.com/dtds/podcast-1.0.dtd"],
	podcast: [
		"https://podcastindex.org/namespace/1.0",
		"http://podcastindex.org/namespace/1.0",
		"https://podcastindex.org/namespace/1.0/",
		"http://podcastindex.org/namespace/1.0/",
		"https://github.com/Podcastindex-org/podcast-namespace/blob/main/docs/1.0.md"
	],
	psc: [
		"http://podlove.org/simple-chapters",
		"https://podlove.org/simple-chapters",
		"http://podlove.org/simple-chapters/",
		"https://podlove.org/simple-chapters/"
	],
	media: [
		"http://search.yahoo.com/mrss/",
		"https://search.yahoo.com/mrss/",
		"http://search.yahoo.com/mrss",
		"https://search.yahoo.com/mrss",
		"http://video.search.yahoo.com/mrss",
		"http://video.search.yahoo.com/mrss/",
		"https://video.search.yahoo.com/mrss",
		"https://video.search.yahoo.com/mrss/",
		"http://www.rssboard.org/media-rss",
		"http://www.rssboard.org/media-rss/",
		"https://www.rssboard.org/media-rss",
		"https://www.rssboard.org/media-rss/",
		"http://search.yahoo.com/searchmonkey/media/",
		"https://search.yahoo.com/searchmonkey/media/",
		"http://search.yahoo.com/searchmonkey/media",
		"https://search.yahoo.com/searchmonkey/media",
		"http://tools.search.yahoo.com/mrss/",
		"https://tools.search.yahoo.com/mrss/",
		"http://tools.search.yahoo.com/mrss",
		"https://tools.search.yahoo.com/mrss"
	],
	georss: [
		"http://www.georss.org/georss",
		"http://www.georss.org/georss/",
		"https://www.georss.org/georss",
		"https://www.georss.org/georss/"
	],
	geo: ["http://www.w3.org/2003/01/geo/wgs84_pos#", "https://www.w3.org/2003/01/geo/wgs84_pos#"],
	thr: [
		"http://purl.org/syndication/thread/1.0",
		"https://purl.org/syndication/thread/1.0",
		"http://purl.org/syndication/thread/1.0/",
		"https://purl.org/syndication/thread/1.0/",
		"http://purl.org/rss/1.0/modules/threading/",
		"https://purl.org/rss/1.0/modules/threading/",
		"http://purl.org/rss/1.0/modules/threading",
		"https://purl.org/rss/1.0/modules/threading"
	],
	dcterms: [
		"http://purl.org/dc/terms/",
		"https://purl.org/dc/terms/",
		"http://purl.org/dc/terms",
		"https://purl.org/dc/terms",
		"http://dublincore.org/documents/dcmi-terms/",
		"https://dublincore.org/documents/dcmi-terms/",
		"http://dublincore.org/documents/dcmi-terms",
		"https://dublincore.org/documents/dcmi-terms",
		"http://dublincore.org/specifications/dublin-core/dcmi-terms/",
		"https://dublincore.org/specifications/dublin-core/dcmi-terms/",
		"http://dublincore.org/specifications/dublin-core/dcmi-terms",
		"https://dublincore.org/specifications/dublin-core/dcmi-terms"
	],
	wfw: [
		"http://wellformedweb.org/CommentAPI/",
		"https://wellformedweb.org/CommentAPI/",
		"http://wellformedweb.org/CommentAPI",
		"https://wellformedweb.org/CommentAPI"
	],
	source: [
		"http://source.scripting.com/",
		"https://source.scripting.com/",
		"http://source.scripting.com",
		"https://source.scripting.com"
	],
	feedpress: [
		"https://feed.press/xmlns",
		"http://feed.press/xmlns",
		"https://feed.press/xmlns/",
		"http://feed.press/xmlns/",
		"https://feedpress.com/xmlns",
		"http://feedpress.com/xmlns",
		"https://feedpress.com/xmlns/",
		"http://feedpress.com/xmlns/",
		"http://feedpress.it/xmlns",
		"https://feedpress.it/xmlns",
		"http://feedpress.it/xmlns/",
		"https://feedpress.it/xmlns/"
	],
	yt: [
		"http://www.youtube.com/xml/schemas/2015",
		"https://www.youtube.com/xml/schemas/2015",
		"http://www.youtube.com/xml/schemas/2015/",
		"https://www.youtube.com/xml/schemas/2015/"
	],
	googleplay: [
		"https://www.google.com/schemas/play-podcasts/1.0/",
		"http://www.google.com/schemas/play-podcasts/1.0/",
		"https://www.google.com/schemas/play-podcasts/1.0",
		"http://www.google.com/schemas/play-podcasts/1.0",
		"https://www.google.com/schemas/play-podcasts/1.0/play-podcasts.xsd",
		"http://www.google.com/schemas/play-podcasts/1.0/play-podcasts.xsd"
	],
	spotify: [
		"http://www.spotify.com/ns/rss",
		"https://www.spotify.com/ns/rss",
		"http://www.spotify.com/ns/rss/",
		"https://www.spotify.com/ns/rss/"
	],
	rdf: ["http://www.w3.org/1999/02/22-rdf-syntax-ns#", "https://www.w3.org/1999/02/22-rdf-syntax-ns#"],
	rawvoice: [
		"http://www.rawvoice.com/rawvoiceRssModule/",
		"https://www.rawvoice.com/rawvoiceRssModule/",
		"http://www.rawvoice.com/rawvoiceRssModule",
		"https://www.rawvoice.com/rawvoiceRssModule",
		"https://blubrry.com/developer/rawvoice-rss",
		"http://blubrry.com/developer/rawvoice-rss",
		"https://blubrry.com/developer/rawvoice-rss/",
		"http://blubrry.com/developer/rawvoice-rss/"
	],
	cc: [
		"http://creativecommons.org/ns#",
		"https://creativecommons.org/ns#",
		"http://web.resource.org/cc/",
		"https://web.resource.org/cc/",
		"http://web.resource.org/cc",
		"https://web.resource.org/cc"
	],
	opensearch: [
		"http://a9.com/-/spec/opensearch/1.1/",
		"https://a9.com/-/spec/opensearch/1.1/",
		"http://a9.com/-/spec/opensearch/1.0/",
		"https://a9.com/-/spec/opensearch/1.0/",
		"http://a9.com/-/spec/opensearchrss/1.0/",
		"https://a9.com/-/spec/opensearchrss/1.0/"
	],
	arxiv: [
		"http://arxiv.org/schemas/atom",
		"https://arxiv.org/schemas/atom",
		"http://arxiv.org/schemas/atom/",
		"https://arxiv.org/schemas/atom/"
	],
	pingback: [
		"http://madskills.com/public/xml/rss/module/pingback/",
		"https://madskills.com/public/xml/rss/module/pingback/",
		"http://madskills.com/public/xml/rss/module/pingback",
		"https://madskills.com/public/xml/rss/module/pingback"
	],
	trackback: [
		"http://madskills.com/public/xml/rss/module/trackback/",
		"https://madskills.com/public/xml/rss/module/trackback/",
		"http://madskills.com/public/xml/rss/module/trackback",
		"https://madskills.com/public/xml/rss/module/trackback"
	],
	prism: [
		"http://prismstandard.org/namespaces/basic/3.0/",
		"https://prismstandard.org/namespaces/basic/3.0/",
		"http://prismstandard.org/namespaces/basic/3.0",
		"https://prismstandard.org/namespaces/basic/3.0",
		"http://prismstandard.org/namespaces/basic/2.2/",
		"https://prismstandard.org/namespaces/basic/2.2/",
		"http://prismstandard.org/namespaces/basic/2.2",
		"https://prismstandard.org/namespaces/basic/2.2",
		"http://prismstandard.org/namespaces/basic/2.1/",
		"https://prismstandard.org/namespaces/basic/2.1/",
		"http://prismstandard.org/namespaces/basic/2.1",
		"https://prismstandard.org/namespaces/basic/2.1",
		"http://prismstandard.org/namespaces/basic/2.0/",
		"https://prismstandard.org/namespaces/basic/2.0/",
		"http://prismstandard.org/namespaces/basic/2.0",
		"https://prismstandard.org/namespaces/basic/2.0",
		"http://prismstandard.org/namespaces/2.0/basic/",
		"https://prismstandard.org/namespaces/2.0/basic/",
		"http://prismstandard.org/namespaces/2.0/basic",
		"https://prismstandard.org/namespaces/2.0/basic",
		"http://prismstandard.org/namespaces/1.2/basic/",
		"https://prismstandard.org/namespaces/1.2/basic/",
		"http://prismstandard.org/namespaces/1.2/basic",
		"https://prismstandard.org/namespaces/1.2/basic",
		"http://prismstandard.org/namespaces/1.1/basic/",
		"https://prismstandard.org/namespaces/1.1/basic/",
		"http://prismstandard.org/namespaces/1.1/basic",
		"https://prismstandard.org/namespaces/1.1/basic",
		"http://prismstandard.org/namespaces/1.0/basic/",
		"https://prismstandard.org/namespaces/1.0/basic/",
		"http://prismstandard.org/namespaces/1.0/basic",
		"https://prismstandard.org/namespaces/1.0/basic",
		"http://purl.org/rss/1.0/modules/prism/",
		"https://purl.org/rss/1.0/modules/prism/",
		"http://purl.org/rss/1.0/modules/prism",
		"https://purl.org/rss/1.0/modules/prism"
	],
	acast: [
		"https://schema.acast.com/1.0/",
		"http://schema.acast.com/1.0/",
		"https://schema.acast.com/1.0",
		"http://schema.acast.com/1.0"
	]
};
const namespacePrefixes = Object.entries(namespaceUris).reduce((prefixes, [prefix, uris]) => {
	for (const uri of uris) {
		const normalizedUri = uri.toLowerCase();
		prefixes[normalizedUri] = prefix.toLowerCase();
	}
	return prefixes;
}, {});

//#endregion
exports.builderConfig = builderConfig;
exports.locales = locales;
exports.namespacePrefixes = namespacePrefixes;
exports.namespaceUris = namespaceUris;
exports.parserConfig = parserConfig;
