// feedsmith/dist/namespaces/thr/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/thr/parse/utils.ts
const parseInReplyTo = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		ref: require_utils.parseString(value["@ref"]),
		href: require_utils.parseString(value["@href"]),
		type: require_utils.parseString(value["@type"]),
		source: require_utils.parseString(value["@source"])
	});
};
const retrieveLink = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		count: require_utils.parseNumber(value["@thr:count"]),
		updated: require_utils.parseDate(value["@thr:updated"])
	});
};
const retrieveItem = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		total: require_utils.parseSingularOf(value["thr:total"], (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		inReplyTos: require_utils.parseArrayOf(value["thr:in-reply-to"], parseInReplyTo)
	});
};

//#endregion
exports.retrieveItem = retrieveItem;
exports.retrieveLink = retrieveLink;
