// feedsmith/dist/namespaces/itunes/parse/utils.js
import { isNonEmptyStringOrNumber, isObject, parseArrayOf, parseBoolean, parseCsvOf, parseNumber, parseSingularOf, parseString, parseYesNoBoolean, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/itunes/parse/utils.ts
const parseCategory = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		text: parseString(value["@text"]),
		categories: parseArrayOf(value["itunes:category"], parseCategory)
	});
};
const parseOwner = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		name: parseSingularOf(value["itunes:name"], (value$1) => parseString(retrieveText(value$1))),
		email: parseSingularOf(value["itunes:email"], (value$1) => parseString(retrieveText(value$1)))
	});
};
const explicitOrYesRegex = /^\p{White_Space}*(explicit|yes)\p{White_Space}*$/iu;
const parseExplicit = (value) => {
	const boolean = parseBoolean(value);
	if (boolean !== void 0) return boolean;
	if (typeof value === "string") return explicitOrYesRegex.test(value);
};
const parseDuration = (value) => {
	const duration = parseNumber(value);
	if (duration !== void 0) return duration;
	if (typeof value !== "string") return;
	const match = value.match(/^(?:(\d+):)?(\d+):(\d+)$/);
	if (match) {
		const [, hours, minutes, seconds] = match;
		return Number(hours || 0) * 3600 + Number(minutes) * 60 + Number(seconds);
	}
};
const parseImage = (value) => {
	if (isNonEmptyStringOrNumber(value)) return parseString(value);
	if (!isObject(value)) return;
	return parseString(value["@href"]);
};
const retrieveItem = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		duration: parseSingularOf(value["itunes:duration"], (value$1) => parseDuration(retrieveText(value$1))),
		image: parseSingularOf(value["itunes:image"], parseImage),
		explicit: parseSingularOf(value["itunes:explicit"], (value$1) => parseExplicit(retrieveText(value$1))),
		author: parseSingularOf(value["itunes:author"], (value$1) => parseString(retrieveText(value$1))),
		title: parseSingularOf(value["itunes:title"], (value$1) => parseString(retrieveText(value$1))),
		episode: parseSingularOf(value["itunes:episode"], (value$1) => parseNumber(retrieveText(value$1))),
		season: parseSingularOf(value["itunes:season"], (value$1) => parseNumber(retrieveText(value$1))),
		episodeType: parseSingularOf(value["itunes:episodetype"], (value$1) => parseString(retrieveText(value$1))),
		block: parseSingularOf(value["itunes:block"], (value$1) => parseYesNoBoolean(retrieveText(value$1))),
		summary: parseSingularOf(value["itunes:summary"], (value$1) => parseString(retrieveText(value$1))),
		subtitle: parseSingularOf(value["itunes:subtitle"], (value$1) => parseString(retrieveText(value$1))),
		keywords: parseSingularOf(value["itunes:keywords"], (value$1) => parseCsvOf(retrieveText(value$1), parseString))
	});
};
const retrieveFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		image: parseSingularOf(value["itunes:image"], parseImage),
		explicit: parseSingularOf(value["itunes:explicit"], (value$1) => parseExplicit(retrieveText(value$1))),
		author: parseSingularOf(value["itunes:author"], (value$1) => parseString(retrieveText(value$1))),
		title: parseSingularOf(value["itunes:title"], (value$1) => parseString(retrieveText(value$1))),
		type: parseSingularOf(value["itunes:type"], (value$1) => parseString(retrieveText(value$1))),
		newFeedUrl: parseSingularOf(value["itunes:new-feed-url"], (value$1) => parseString(retrieveText(value$1))),
		block: parseSingularOf(value["itunes:block"], (value$1) => parseYesNoBoolean(retrieveText(value$1))),
		complete: parseSingularOf(value["itunes:complete"], (value$1) => parseYesNoBoolean(retrieveText(value$1))),
		applePodcastsVerify: parseSingularOf(value["itunes:applepodcastsverify"], (value$1) => parseString(retrieveText(value$1))),
		categories: parseArrayOf(value["itunes:category"], parseCategory),
		owner: parseSingularOf(value["itunes:owner"], parseOwner),
		summary: parseSingularOf(value["itunes:summary"], (value$1) => parseString(retrieveText(value$1))),
		subtitle: parseSingularOf(value["itunes:subtitle"], (value$1) => parseString(retrieveText(value$1))),
		keywords: parseSingularOf(value["itunes:keywords"], (value$1) => parseCsvOf(retrieveText(value$1), parseString))
	});
};

//#endregion
export { retrieveFeed, retrieveItem };
