// feedsmith/dist/namespaces/itunes/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/itunes/parse/utils.ts
const parseCategory = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		text: require_utils.parseString(value["@text"]),
		categories: require_utils.parseArrayOf(value["itunes:category"], parseCategory)
	});
};
const parseOwner = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		name: require_utils.parseSingularOf(value["itunes:name"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		email: require_utils.parseSingularOf(value["itunes:email"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};
const explicitOrYesRegex = /^\p{White_Space}*(explicit|yes)\p{White_Space}*$/iu;
const parseExplicit = (value) => {
	const boolean = require_utils.parseBoolean(value);
	if (boolean !== void 0) return boolean;
	if (typeof value === "string") return explicitOrYesRegex.test(value);
};
const parseDuration = (value) => {
	const duration = require_utils.parseNumber(value);
	if (duration !== void 0) return duration;
	if (typeof value !== "string") return;
	const match = value.match(/^(?:(\d+):)?(\d+):(\d+)$/);
	if (match) {
		const [, hours, minutes, seconds] = match;
		return Number(hours || 0) * 3600 + Number(minutes) * 60 + Number(seconds);
	}
};
const parseImage = (value) => {
	if (require_utils.isNonEmptyStringOrNumber(value)) return require_utils.parseString(value);
	if (!require_utils.isObject(value)) return;
	return require_utils.parseString(value["@href"]);
};
const retrieveItem = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		duration: require_utils.parseSingularOf(value["itunes:duration"], (value$1) => parseDuration(require_utils.retrieveText(value$1))),
		image: require_utils.parseSingularOf(value["itunes:image"], parseImage),
		explicit: require_utils.parseSingularOf(value["itunes:explicit"], (value$1) => parseExplicit(require_utils.retrieveText(value$1))),
		author: require_utils.parseSingularOf(value["itunes:author"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		title: require_utils.parseSingularOf(value["itunes:title"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		episode: require_utils.parseSingularOf(value["itunes:episode"], (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		season: require_utils.parseSingularOf(value["itunes:season"], (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		episodeType: require_utils.parseSingularOf(value["itunes:episodetype"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		block: require_utils.parseSingularOf(value["itunes:block"], (value$1) => require_utils.parseYesNoBoolean(require_utils.retrieveText(value$1))),
		summary: require_utils.parseSingularOf(value["itunes:summary"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		subtitle: require_utils.parseSingularOf(value["itunes:subtitle"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		keywords: require_utils.parseSingularOf(value["itunes:keywords"], (value$1) => require_utils.parseCsvOf(require_utils.retrieveText(value$1), require_utils.parseString))
	});
};
const retrieveFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		image: require_utils.parseSingularOf(value["itunes:image"], parseImage),
		explicit: require_utils.parseSingularOf(value["itunes:explicit"], (value$1) => parseExplicit(require_utils.retrieveText(value$1))),
		author: require_utils.parseSingularOf(value["itunes:author"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		title: require_utils.parseSingularOf(value["itunes:title"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		type: require_utils.parseSingularOf(value["itunes:type"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		newFeedUrl: require_utils.parseSingularOf(value["itunes:new-feed-url"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		block: require_utils.parseSingularOf(value["itunes:block"], (value$1) => require_utils.parseYesNoBoolean(require_utils.retrieveText(value$1))),
		complete: require_utils.parseSingularOf(value["itunes:complete"], (value$1) => require_utils.parseYesNoBoolean(require_utils.retrieveText(value$1))),
		applePodcastsVerify: require_utils.parseSingularOf(value["itunes:applepodcastsverify"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		categories: require_utils.parseArrayOf(value["itunes:category"], parseCategory),
		owner: require_utils.parseSingularOf(value["itunes:owner"], parseOwner),
		summary: require_utils.parseSingularOf(value["itunes:summary"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		subtitle: require_utils.parseSingularOf(value["itunes:subtitle"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		keywords: require_utils.parseSingularOf(value["itunes:keywords"], (value$1) => require_utils.parseCsvOf(require_utils.retrieveText(value$1), require_utils.parseString))
	});
};

//#endregion
exports.retrieveFeed = retrieveFeed;
exports.retrieveItem = retrieveItem;
