// feedsmith/dist/namespaces/opensearch/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/opensearch/parse/utils.ts
const parseQuery = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		role: require_utils.parseString(value["@role"]),
		searchTerms: require_utils.parseString(value["@searchterms"]),
		count: require_utils.parseNumber(value["@count"]),
		startIndex: require_utils.parseNumber(value["@startindex"]),
		startPage: require_utils.parseNumber(value["@startpage"]),
		language: require_utils.parseString(value["@language"]),
		inputEncoding: require_utils.parseString(value["@inputencoding"]),
		outputEncoding: require_utils.parseString(value["@outputencoding"])
	});
};
const retrieveFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		totalResults: require_utils.parseSingularOf(value["opensearch:totalresults"], (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		startIndex: require_utils.parseSingularOf(value["opensearch:startindex"], (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		itemsPerPage: require_utils.parseSingularOf(value["opensearch:itemsperpage"], (value$1) => require_utils.parseNumber(require_utils.retrieveText(value$1))),
		queries: require_utils.parseArrayOf(value["opensearch:query"], parseQuery)
	});
};

//#endregion
exports.retrieveFeed = retrieveFeed;
