// feedsmith/dist/namespaces/opensearch/parse/utils.js
import { isObject, parseArrayOf, parseNumber, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/opensearch/parse/utils.ts
const parseQuery = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		role: parseString(value["@role"]),
		searchTerms: parseString(value["@searchterms"]),
		count: parseNumber(value["@count"]),
		startIndex: parseNumber(value["@startindex"]),
		startPage: parseNumber(value["@startpage"]),
		language: parseString(value["@language"]),
		inputEncoding: parseString(value["@inputencoding"]),
		outputEncoding: parseString(value["@outputencoding"])
	});
};
const retrieveFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		totalResults: parseSingularOf(value["opensearch:totalresults"], (value$1) => parseNumber(retrieveText(value$1))),
		startIndex: parseSingularOf(value["opensearch:startindex"], (value$1) => parseNumber(retrieveText(value$1))),
		itemsPerPage: parseSingularOf(value["opensearch:itemsperpage"], (value$1) => parseNumber(retrieveText(value$1))),
		queries: parseArrayOf(value["opensearch:query"], parseQuery)
	});
};

//#endregion
export { retrieveFeed };
