// feedsmith/dist/namespaces/googleplay/parse/utils.js
import { isObject, parseArrayOf, parseSingularOf, parseString, parseYesNoBoolean, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/googleplay/parse/utils.ts
const parseImage = (value) => {
	if (isObject(value) && value["@href"]) return trimObject({ href: parseString(value["@href"]) });
	const text = retrieveText(value);
	if (text) return trimObject({ href: parseString(text) });
};
const parseCategory = (value) => {
	if (isObject(value) && value["@text"]) return parseString(value["@text"]);
	return parseString(retrieveText(value));
};
const parseExplicit = (value) => {
	const explicit = retrieveText(value)?.trim().toLowerCase();
	if (explicit === "clean") return explicit;
	return parseYesNoBoolean(value);
};
const retrieveItem = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		author: parseSingularOf(value["googleplay:author"], (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value["googleplay:description"], (value$1) => parseString(retrieveText(value$1))),
		explicit: parseSingularOf(value["googleplay:explicit"], parseExplicit),
		block: parseSingularOf(value["googleplay:block"], (value$1) => parseYesNoBoolean(retrieveText(value$1))),
		image: parseSingularOf(value["googleplay:image"], parseImage)
	});
};
const retrieveFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		author: parseSingularOf(value["googleplay:author"], (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value["googleplay:description"], (value$1) => parseString(retrieveText(value$1))),
		explicit: parseSingularOf(value["googleplay:explicit"], parseExplicit),
		block: parseSingularOf(value["googleplay:block"], (value$1) => parseYesNoBoolean(retrieveText(value$1))),
		image: parseSingularOf(value["googleplay:image"], parseImage),
		newFeedUrl: parseSingularOf(value["googleplay:new-feed-url"], (value$1) => parseString(retrieveText(value$1))),
		email: parseSingularOf(value["googleplay:email"], (value$1) => parseString(retrieveText(value$1))),
		categories: parseArrayOf(value["googleplay:category"], parseCategory)
	});
};

//#endregion
export { retrieveFeed, retrieveItem };
