// feedsmith/dist/namespaces/googleplay/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/googleplay/parse/utils.ts
const parseImage = (value) => {
	if (require_utils.isObject(value) && value["@href"]) return require_utils.trimObject({ href: require_utils.parseString(value["@href"]) });
	const text = require_utils.retrieveText(value);
	if (text) return require_utils.trimObject({ href: require_utils.parseString(text) });
};
const parseCategory = (value) => {
	if (require_utils.isObject(value) && value["@text"]) return require_utils.parseString(value["@text"]);
	return require_utils.parseString(require_utils.retrieveText(value));
};
const parseExplicit = (value) => {
	const explicit = require_utils.retrieveText(value)?.trim().toLowerCase();
	if (explicit === "clean") return explicit;
	return require_utils.parseYesNoBoolean(value);
};
const retrieveItem = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		author: require_utils.parseSingularOf(value["googleplay:author"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		description: require_utils.parseSingularOf(value["googleplay:description"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		explicit: require_utils.parseSingularOf(value["googleplay:explicit"], parseExplicit),
		block: require_utils.parseSingularOf(value["googleplay:block"], (value$1) => require_utils.parseYesNoBoolean(require_utils.retrieveText(value$1))),
		image: require_utils.parseSingularOf(value["googleplay:image"], parseImage)
	});
};
const retrieveFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		author: require_utils.parseSingularOf(value["googleplay:author"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		description: require_utils.parseSingularOf(value["googleplay:description"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		explicit: require_utils.parseSingularOf(value["googleplay:explicit"], parseExplicit),
		block: require_utils.parseSingularOf(value["googleplay:block"], (value$1) => require_utils.parseYesNoBoolean(require_utils.retrieveText(value$1))),
		image: require_utils.parseSingularOf(value["googleplay:image"], parseImage),
		newFeedUrl: require_utils.parseSingularOf(value["googleplay:new-feed-url"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		email: require_utils.parseSingularOf(value["googleplay:email"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		categories: require_utils.parseArrayOf(value["googleplay:category"], parseCategory)
	});
};

//#endregion
exports.retrieveFeed = retrieveFeed;
exports.retrieveItem = retrieveItem;
