// feedsmith/dist/namespaces/source/parse/utils.js
import { isObject, parseArrayOf, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/source/parse/utils.ts
const parseAccount = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		service: parseString(value["@service"]),
		value: parseString(retrieveText(value))
	});
};
const parseLikes = (value) => {
	if (!isObject(value)) return;
	return trimObject({ server: parseString(value["@server"]) });
};
const parseArchive = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		url: parseSingularOf(value["source:url"], (value$1) => parseString(retrieveText(value$1))),
		startDay: parseSingularOf(value["source:startday"], (value$1) => parseString(retrieveText(value$1))),
		endDay: parseSingularOf(value["source:endday"], (value$1) => parseString(retrieveText(value$1))),
		filename: parseSingularOf(value["source:filename"], (value$1) => parseString(retrieveText(value$1)))
	});
};
const parseSubscriptionList = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		url: parseString(value["@url"]),
		value: parseString(retrieveText(value))
	});
};
const retrieveFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		accounts: parseArrayOf(value["source:account"], parseAccount),
		likes: parseSingularOf(value["source:likes"], parseLikes),
		archive: parseSingularOf(value["source:archive"], parseArchive),
		subscriptionLists: parseArrayOf(value["source:subscriptionlist"], parseSubscriptionList),
		cloud: parseSingularOf(value["source:cloud"], (value$1) => parseString(retrieveText(value$1))),
		blogroll: parseSingularOf(value["source:blogroll"], (value$1) => parseString(retrieveText(value$1))),
		self: parseSingularOf(value["source:self"], (value$1) => parseString(retrieveText(value$1)))
	});
};
const retrieveItem = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		markdown: parseSingularOf(value["source:markdown"], (value$1) => parseString(retrieveText(value$1))),
		outlines: parseArrayOf(value["source:outline"], (value$1) => parseString(retrieveText(value$1))),
		localTime: parseSingularOf(value["source:localtime"], (value$1) => parseString(retrieveText(value$1))),
		linkFull: parseSingularOf(value["source:linkfull"], (value$1) => parseString(retrieveText(value$1)))
	});
};

//#endregion
export { retrieveFeed, retrieveItem };
