// feedsmith/dist/namespaces/source/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/source/parse/utils.ts
const parseAccount = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		service: require_utils.parseString(value["@service"]),
		value: require_utils.parseString(require_utils.retrieveText(value))
	});
};
const parseLikes = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({ server: require_utils.parseString(value["@server"]) });
};
const parseArchive = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		url: require_utils.parseSingularOf(value["source:url"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		startDay: require_utils.parseSingularOf(value["source:startday"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		endDay: require_utils.parseSingularOf(value["source:endday"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		filename: require_utils.parseSingularOf(value["source:filename"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};
const parseSubscriptionList = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		url: require_utils.parseString(value["@url"]),
		value: require_utils.parseString(require_utils.retrieveText(value))
	});
};
const retrieveFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		accounts: require_utils.parseArrayOf(value["source:account"], parseAccount),
		likes: require_utils.parseSingularOf(value["source:likes"], parseLikes),
		archive: require_utils.parseSingularOf(value["source:archive"], parseArchive),
		subscriptionLists: require_utils.parseArrayOf(value["source:subscriptionlist"], parseSubscriptionList),
		cloud: require_utils.parseSingularOf(value["source:cloud"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		blogroll: require_utils.parseSingularOf(value["source:blogroll"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		self: require_utils.parseSingularOf(value["source:self"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};
const retrieveItem = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		markdown: require_utils.parseSingularOf(value["source:markdown"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		outlines: require_utils.parseArrayOf(value["source:outline"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		localTime: require_utils.parseSingularOf(value["source:localtime"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		linkFull: require_utils.parseSingularOf(value["source:linkfull"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};

//#endregion
exports.retrieveFeed = retrieveFeed;
exports.retrieveItem = retrieveItem;
