// feedsmith/dist/namespaces/cc/parse/utils.js
import { isObject, parseSingularOf, parseString, retrieveRdfResourceOrText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/cc/parse/utils.ts
const retrieveItemOrFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		license: parseSingularOf(value["cc:license"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		morePermissions: parseSingularOf(value["cc:morepermissions"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		attributionName: parseSingularOf(value["cc:attributionname"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		attributionURL: parseSingularOf(value["cc:attributionurl"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		useGuidelines: parseSingularOf(value["cc:useguidelines"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		permits: parseSingularOf(value["cc:permits"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		requires: parseSingularOf(value["cc:requires"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		prohibits: parseSingularOf(value["cc:prohibits"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		jurisdiction: parseSingularOf(value["cc:jurisdiction"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		legalcode: parseSingularOf(value["cc:legalcode"], (value$1) => retrieveRdfResourceOrText(value$1, parseString)),
		deprecatedOn: parseSingularOf(value["cc:deprecatedon"], (value$1) => retrieveRdfResourceOrText(value$1, parseString))
	});
};

//#endregion
export { retrieveItemOrFeed };
