// feedsmith/dist/namespaces/cc/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/cc/parse/utils.ts
const retrieveItemOrFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		license: require_utils.parseSingularOf(value["cc:license"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		morePermissions: require_utils.parseSingularOf(value["cc:morepermissions"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		attributionName: require_utils.parseSingularOf(value["cc:attributionname"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		attributionURL: require_utils.parseSingularOf(value["cc:attributionurl"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		useGuidelines: require_utils.parseSingularOf(value["cc:useguidelines"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		permits: require_utils.parseSingularOf(value["cc:permits"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		requires: require_utils.parseSingularOf(value["cc:requires"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		prohibits: require_utils.parseSingularOf(value["cc:prohibits"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		jurisdiction: require_utils.parseSingularOf(value["cc:jurisdiction"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		legalcode: require_utils.parseSingularOf(value["cc:legalcode"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString)),
		deprecatedOn: require_utils.parseSingularOf(value["cc:deprecatedon"], (value$1) => require_utils.retrieveRdfResourceOrText(value$1, require_utils.parseString))
	});
};

//#endregion
exports.retrieveItemOrFeed = retrieveItemOrFeed;
