// feedsmith/dist/namespaces/dcterms/parse/utils.js
import { isObject, parseArrayOf, parseDate, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/dcterms/parse/utils.ts
const retrieveItemOrFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		abstracts: parseArrayOf(value["dcterms:abstract"], (value$1) => parseString(retrieveText(value$1))),
		accrualMethods: parseArrayOf(value["dcterms:accrualmethod"], (value$1) => parseString(retrieveText(value$1))),
		accrualPeriodicities: parseArrayOf(value["dcterms:accrualperiodicity"], (value$1) => parseString(retrieveText(value$1))),
		accrualPolicies: parseArrayOf(value["dcterms:accrualpolicy"], (value$1) => parseString(retrieveText(value$1))),
		alternatives: parseArrayOf(value["dcterms:alternative"], (value$1) => parseString(retrieveText(value$1))),
		audiences: parseArrayOf(value["dcterms:audience"], (value$1) => parseString(retrieveText(value$1))),
		bibliographicCitations: parseArrayOf(value["dcterms:bibliographiccitation"], (value$1) => parseString(retrieveText(value$1))),
		contributors: parseArrayOf(value["dcterms:contributor"], (value$1) => parseString(retrieveText(value$1))),
		coverages: parseArrayOf(value["dcterms:coverage"], (value$1) => parseString(retrieveText(value$1))),
		creators: parseArrayOf(value["dcterms:creator"], (value$1) => parseString(retrieveText(value$1))),
		dates: parseArrayOf(value["dcterms:date"], (value$1) => parseDate(retrieveText(value$1))),
		descriptions: parseArrayOf(value["dcterms:description"], (value$1) => parseString(retrieveText(value$1))),
		educationLevels: parseArrayOf(value["dcterms:educationlevel"], (value$1) => parseString(retrieveText(value$1))),
		extents: parseArrayOf(value["dcterms:extent"], (value$1) => parseString(retrieveText(value$1))),
		formats: parseArrayOf(value["dcterms:format"], (value$1) => parseString(retrieveText(value$1))),
		hasFormats: parseArrayOf(value["dcterms:hasformat"], (value$1) => parseString(retrieveText(value$1))),
		hasParts: parseArrayOf(value["dcterms:haspart"], (value$1) => parseString(retrieveText(value$1))),
		hasVersions: parseArrayOf(value["dcterms:hasversion"], (value$1) => parseString(retrieveText(value$1))),
		identifiers: parseArrayOf(value["dcterms:identifier"], (value$1) => parseString(retrieveText(value$1))),
		instructionalMethods: parseArrayOf(value["dcterms:instructionalmethod"], (value$1) => parseString(retrieveText(value$1))),
		languages: parseArrayOf(value["dcterms:language"], (value$1) => parseString(retrieveText(value$1))),
		licenses: parseArrayOf(value["dcterms:license"], (value$1) => parseString(retrieveText(value$1))),
		mediators: parseArrayOf(value["dcterms:mediator"], (value$1) => parseString(retrieveText(value$1))),
		mediums: parseArrayOf(value["dcterms:medium"], (value$1) => parseString(retrieveText(value$1))),
		provenances: parseArrayOf(value["dcterms:provenance"], (value$1) => parseString(retrieveText(value$1))),
		publishers: parseArrayOf(value["dcterms:publisher"], (value$1) => parseString(retrieveText(value$1))),
		relations: parseArrayOf(value["dcterms:relation"], (value$1) => parseString(retrieveText(value$1))),
		rightsHolders: parseArrayOf(value["dcterms:rightsholder"], (value$1) => parseString(retrieveText(value$1))),
		sources: parseArrayOf(value["dcterms:source"], (value$1) => parseString(retrieveText(value$1))),
		spatials: parseArrayOf(value["dcterms:spatial"], (value$1) => parseString(retrieveText(value$1))),
		subjects: parseArrayOf(value["dcterms:subject"], (value$1) => parseString(retrieveText(value$1))),
		temporals: parseArrayOf(value["dcterms:temporal"], (value$1) => parseString(retrieveText(value$1))),
		titles: parseArrayOf(value["dcterms:title"], (value$1) => parseString(retrieveText(value$1))),
		types: parseArrayOf(value["dcterms:type"], (value$1) => parseString(retrieveText(value$1))),
		abstract: parseSingularOf(value["dcterms:abstract"], (value$1) => parseString(retrieveText(value$1))),
		accrualMethod: parseSingularOf(value["dcterms:accrualmethod"], (value$1) => parseString(retrieveText(value$1))),
		accrualPeriodicity: parseSingularOf(value["dcterms:accrualperiodicity"], (value$1) => parseString(retrieveText(value$1))),
		accrualPolicy: parseSingularOf(value["dcterms:accrualpolicy"], (value$1) => parseString(retrieveText(value$1))),
		alternative: parseSingularOf(value["dcterms:alternative"], (value$1) => parseString(retrieveText(value$1))),
		audience: parseSingularOf(value["dcterms:audience"], (value$1) => parseString(retrieveText(value$1))),
		bibliographicCitation: parseSingularOf(value["dcterms:bibliographiccitation"], (value$1) => parseString(retrieveText(value$1))),
		contributor: parseSingularOf(value["dcterms:contributor"], (value$1) => parseString(retrieveText(value$1))),
		coverage: parseSingularOf(value["dcterms:coverage"], (value$1) => parseString(retrieveText(value$1))),
		creator: parseSingularOf(value["dcterms:creator"], (value$1) => parseString(retrieveText(value$1))),
		date: parseSingularOf(value["dcterms:date"], (value$1) => parseDate(retrieveText(value$1))),
		description: parseSingularOf(value["dcterms:description"], (value$1) => parseString(retrieveText(value$1))),
		educationLevel: parseSingularOf(value["dcterms:educationlevel"], (value$1) => parseString(retrieveText(value$1))),
		extent: parseSingularOf(value["dcterms:extent"], (value$1) => parseString(retrieveText(value$1))),
		format: parseSingularOf(value["dcterms:format"], (value$1) => parseString(retrieveText(value$1))),
		hasFormat: parseSingularOf(value["dcterms:hasformat"], (value$1) => parseString(retrieveText(value$1))),
		hasPart: parseSingularOf(value["dcterms:haspart"], (value$1) => parseString(retrieveText(value$1))),
		hasVersion: parseSingularOf(value["dcterms:hasversion"], (value$1) => parseString(retrieveText(value$1))),
		identifier: parseSingularOf(value["dcterms:identifier"], (value$1) => parseString(retrieveText(value$1))),
		instructionalMethod: parseSingularOf(value["dcterms:instructionalmethod"], (value$1) => parseString(retrieveText(value$1))),
		language: parseSingularOf(value["dcterms:language"], (value$1) => parseString(retrieveText(value$1))),
		license: parseSingularOf(value["dcterms:license"], (value$1) => parseString(retrieveText(value$1))),
		mediator: parseSingularOf(value["dcterms:mediator"], (value$1) => parseString(retrieveText(value$1))),
		medium: parseSingularOf(value["dcterms:medium"], (value$1) => parseString(retrieveText(value$1))),
		provenance: parseSingularOf(value["dcterms:provenance"], (value$1) => parseString(retrieveText(value$1))),
		publisher: parseSingularOf(value["dcterms:publisher"], (value$1) => parseString(retrieveText(value$1))),
		relation: parseSingularOf(value["dcterms:relation"], (value$1) => parseString(retrieveText(value$1))),
		rightsHolder: parseSingularOf(value["dcterms:rightsholder"], (value$1) => parseString(retrieveText(value$1))),
		source: parseSingularOf(value["dcterms:source"], (value$1) => parseString(retrieveText(value$1))),
		spatial: parseSingularOf(value["dcterms:spatial"], (value$1) => parseString(retrieveText(value$1))),
		subject: parseSingularOf(value["dcterms:subject"], (value$1) => parseString(retrieveText(value$1))),
		temporal: parseSingularOf(value["dcterms:temporal"], (value$1) => parseString(retrieveText(value$1))),
		title: parseSingularOf(value["dcterms:title"], (value$1) => parseString(retrieveText(value$1))),
		type: parseSingularOf(value["dcterms:type"], (value$1) => parseString(retrieveText(value$1))),
		accessRights: parseSingularOf(value["dcterms:accessrights"], (value$1) => parseString(retrieveText(value$1))),
		available: parseSingularOf(value["dcterms:available"], (value$1) => parseDate(retrieveText(value$1))),
		conformsTo: parseSingularOf(value["dcterms:conformsto"], (value$1) => parseString(retrieveText(value$1))),
		created: parseSingularOf(value["dcterms:created"], (value$1) => parseDate(retrieveText(value$1))),
		dateAccepted: parseSingularOf(value["dcterms:dateaccepted"], (value$1) => parseDate(retrieveText(value$1))),
		dateCopyrighted: parseSingularOf(value["dcterms:datecopyrighted"], (value$1) => parseDate(retrieveText(value$1))),
		dateSubmitted: parseSingularOf(value["dcterms:datesubmitted"], (value$1) => parseDate(retrieveText(value$1))),
		isFormatOf: parseSingularOf(value["dcterms:isformatof"], (value$1) => parseString(retrieveText(value$1))),
		isPartOf: parseSingularOf(value["dcterms:ispartof"], (value$1) => parseString(retrieveText(value$1))),
		isReferencedBy: parseSingularOf(value["dcterms:isreferencedby"], (value$1) => parseString(retrieveText(value$1))),
		isReplacedBy: parseSingularOf(value["dcterms:isreplacedby"], (value$1) => parseString(retrieveText(value$1))),
		isRequiredBy: parseSingularOf(value["dcterms:isrequiredby"], (value$1) => parseString(retrieveText(value$1))),
		issued: parseSingularOf(value["dcterms:issued"], (value$1) => parseDate(retrieveText(value$1))),
		isVersionOf: parseSingularOf(value["dcterms:isversionof"], (value$1) => parseString(retrieveText(value$1))),
		modified: parseSingularOf(value["dcterms:modified"], (value$1) => parseDate(retrieveText(value$1))),
		references: parseSingularOf(value["dcterms:references"], (value$1) => parseString(retrieveText(value$1))),
		replaces: parseSingularOf(value["dcterms:replaces"], (value$1) => parseString(retrieveText(value$1))),
		requires: parseSingularOf(value["dcterms:requires"], (value$1) => parseString(retrieveText(value$1))),
		rights: parseSingularOf(value["dcterms:rights"], (value$1) => parseString(retrieveText(value$1))),
		tableOfContents: parseSingularOf(value["dcterms:tableofcontents"], (value$1) => parseString(retrieveText(value$1))),
		valid: parseSingularOf(value["dcterms:valid"], (value$1) => parseDate(retrieveText(value$1)))
	});
};

//#endregion
export { retrieveItemOrFeed };
