// feedsmith/dist/namespaces/dc/parse/utils.js
import { isObject, parseArrayOf, parseDate, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/dc/parse/utils.ts
const retrieveItemOrFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		titles: parseArrayOf(value["dc:title"], (value$1) => parseString(retrieveText(value$1))),
		creators: parseArrayOf(value["dc:creator"], (value$1) => parseString(retrieveText(value$1))),
		subjects: parseArrayOf(value["dc:subject"], (value$1) => parseString(retrieveText(value$1))),
		descriptions: parseArrayOf(value["dc:description"], (value$1) => parseString(retrieveText(value$1))),
		publishers: parseArrayOf(value["dc:publisher"], (value$1) => parseString(retrieveText(value$1))),
		contributors: parseArrayOf(value["dc:contributor"], (value$1) => parseString(retrieveText(value$1))),
		dates: parseArrayOf(value["dc:date"], (value$1) => parseDate(retrieveText(value$1))),
		types: parseArrayOf(value["dc:type"], (value$1) => parseString(retrieveText(value$1))),
		formats: parseArrayOf(value["dc:format"], (value$1) => parseString(retrieveText(value$1))),
		identifiers: parseArrayOf(value["dc:identifier"], (value$1) => parseString(retrieveText(value$1))),
		sources: parseArrayOf(value["dc:source"], (value$1) => parseString(retrieveText(value$1))),
		languages: parseArrayOf(value["dc:language"], (value$1) => parseString(retrieveText(value$1))),
		relations: parseArrayOf(value["dc:relation"], (value$1) => parseString(retrieveText(value$1))),
		title: parseSingularOf(value["dc:title"], (value$1) => parseString(retrieveText(value$1))),
		creator: parseSingularOf(value["dc:creator"], (value$1) => parseString(retrieveText(value$1))),
		subject: parseSingularOf(value["dc:subject"], (value$1) => parseString(retrieveText(value$1))),
		description: parseSingularOf(value["dc:description"], (value$1) => parseString(retrieveText(value$1))),
		publisher: parseSingularOf(value["dc:publisher"], (value$1) => parseString(retrieveText(value$1))),
		contributor: parseSingularOf(value["dc:contributor"], (value$1) => parseString(retrieveText(value$1))),
		date: parseSingularOf(value["dc:date"], (value$1) => parseDate(retrieveText(value$1))),
		type: parseSingularOf(value["dc:type"], (value$1) => parseString(retrieveText(value$1))),
		format: parseSingularOf(value["dc:format"], (value$1) => parseString(retrieveText(value$1))),
		identifier: parseSingularOf(value["dc:identifier"], (value$1) => parseString(retrieveText(value$1))),
		source: parseSingularOf(value["dc:source"], (value$1) => parseString(retrieveText(value$1))),
		language: parseSingularOf(value["dc:language"], (value$1) => parseString(retrieveText(value$1))),
		relation: parseSingularOf(value["dc:relation"], (value$1) => parseString(retrieveText(value$1))),
		coverage: parseSingularOf(value["dc:coverage"], (value$1) => parseString(retrieveText(value$1))),
		rights: parseSingularOf(value["dc:rights"], (value$1) => parseString(retrieveText(value$1)))
	});
};

//#endregion
export { retrieveItemOrFeed };
