// feedsmith/dist/namespaces/acast/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/acast/parse/utils.ts
const parseSignature = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		key: require_utils.parseString(value["@key"]),
		algorithm: require_utils.parseString(value["@algorithm"]),
		value: require_utils.parseString(require_utils.retrieveText(value))
	});
};
const parseNetwork = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		id: require_utils.parseString(value["@id"]),
		slug: require_utils.parseString(value["@slug"]),
		value: require_utils.parseString(require_utils.retrieveText(value))
	});
};
const retrieveFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		showId: require_utils.parseSingularOf(value["acast:showid"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		showUrl: require_utils.parseSingularOf(value["acast:showurl"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		signature: require_utils.parseSingularOf(value["acast:signature"], parseSignature),
		settings: require_utils.parseSingularOf(value["acast:settings"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		network: require_utils.parseSingularOf(value["acast:network"], parseNetwork),
		importedFeed: require_utils.parseSingularOf(value["acast:importedfeed"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};
const retrieveItem = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		episodeId: require_utils.parseSingularOf(value["acast:episodeid"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		showId: require_utils.parseSingularOf(value["acast:showid"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		episodeUrl: require_utils.parseSingularOf(value["acast:episodeurl"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		settings: require_utils.parseSingularOf(value["acast:settings"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};

//#endregion
exports.retrieveFeed = retrieveFeed;
exports.retrieveItem = retrieveItem;
