// feedsmith/dist/namespaces/acast/parse/utils.js
import { isObject, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/acast/parse/utils.ts
const parseSignature = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		key: parseString(value["@key"]),
		algorithm: parseString(value["@algorithm"]),
		value: parseString(retrieveText(value))
	});
};
const parseNetwork = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		id: parseString(value["@id"]),
		slug: parseString(value["@slug"]),
		value: parseString(retrieveText(value))
	});
};
const retrieveFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		showId: parseSingularOf(value["acast:showid"], (value$1) => parseString(retrieveText(value$1))),
		showUrl: parseSingularOf(value["acast:showurl"], (value$1) => parseString(retrieveText(value$1))),
		signature: parseSingularOf(value["acast:signature"], parseSignature),
		settings: parseSingularOf(value["acast:settings"], (value$1) => parseString(retrieveText(value$1))),
		network: parseSingularOf(value["acast:network"], parseNetwork),
		importedFeed: parseSingularOf(value["acast:importedfeed"], (value$1) => parseString(retrieveText(value$1)))
	});
};
const retrieveItem = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		episodeId: parseSingularOf(value["acast:episodeid"], (value$1) => parseString(retrieveText(value$1))),
		showId: parseSingularOf(value["acast:showid"], (value$1) => parseString(retrieveText(value$1))),
		episodeUrl: parseSingularOf(value["acast:episodeurl"], (value$1) => parseString(retrieveText(value$1))),
		settings: parseSingularOf(value["acast:settings"], (value$1) => parseString(retrieveText(value$1)))
	});
};

//#endregion
export { retrieveFeed, retrieveItem };
