// feedsmith/dist/namespaces/feedpress/parse/utils.cjs
const require_utils = require('../../../common/utils.cjs');

//#region src/namespaces/feedpress/parse/utils.ts
const retrieveFeed = (value) => {
	if (!require_utils.isObject(value)) return;
	return require_utils.trimObject({
		link: require_utils.parseSingularOf(value["feedpress:link"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		newsletterId: require_utils.parseSingularOf(value["feedpress:newsletterid"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		locale: require_utils.parseSingularOf(value["feedpress:locale"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		podcastId: require_utils.parseSingularOf(value["feedpress:podcastid"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1))),
		cssFile: require_utils.parseSingularOf(value["feedpress:cssfile"], (value$1) => require_utils.parseString(require_utils.retrieveText(value$1)))
	});
};

//#endregion
exports.retrieveFeed = retrieveFeed;
