// feedsmith/dist/feeds/json/parse/utils.js
import { isNonEmptyStringOrNumber, isObject, parseArrayOf, parseBoolean, parseDate, parseNumber, parseSingularOf, parseString, trimObject } from "../../../common/utils.js";

//#region src/feeds/json/parse/utils.ts
const createCaseInsensitiveGetter = (value) => {
	return (requestedKey) => {
		if (requestedKey in value) return value[requestedKey];
		const lowerKey = requestedKey.toLowerCase();
		for (const key in value) if (key.toLowerCase() === lowerKey) return value[key];
	};
};
const parseAuthor = (value) => {
	if (isObject(value)) {
		const get = createCaseInsensitiveGetter(value);
		return trimObject({
			name: parseSingularOf(get("name"), parseString),
			url: parseSingularOf(get("url"), parseString),
			avatar: parseSingularOf(get("avatar"), parseString)
		});
	}
	if (isNonEmptyStringOrNumber(value)) return trimObject({ name: parseString(value) });
};
const retrieveAuthors = (value) => {
	if (!isObject(value)) return;
	const get = createCaseInsensitiveGetter(value);
	const parsedAuthors = parseArrayOf(get("authors"), parseAuthor);
	const parsedAuthor = parseArrayOf(get("author"), parseAuthor);
	return parsedAuthors?.length ? parsedAuthors : parsedAuthor;
};
const parseAttachment = (value) => {
	if (!isObject(value)) return;
	const get = createCaseInsensitiveGetter(value);
	return trimObject({
		url: parseSingularOf(get("url"), parseString),
		mime_type: parseSingularOf(get("mime_type"), parseString),
		title: parseSingularOf(get("title"), parseString),
		size_in_bytes: parseSingularOf(get("size_in_bytes"), parseNumber),
		duration_in_seconds: parseSingularOf(get("duration_in_seconds"), parseNumber)
	});
};
const parseItem = (value) => {
	if (!isObject(value)) return;
	const get = createCaseInsensitiveGetter(value);
	return trimObject({
		id: parseSingularOf(get("id"), parseString),
		url: parseSingularOf(get("url"), parseString),
		external_url: parseSingularOf(get("external_url"), parseString),
		title: parseSingularOf(get("title"), parseString),
		content_html: parseSingularOf(get("content_html"), parseString),
		content_text: parseSingularOf(get("content_text"), parseString),
		summary: parseSingularOf(get("summary"), parseString),
		image: parseSingularOf(get("image"), parseString),
		banner_image: parseSingularOf(get("banner_image"), parseString),
		date_published: parseSingularOf(get("date_published"), parseDate),
		date_modified: parseSingularOf(get("date_modified"), parseDate),
		tags: parseArrayOf(get("tags"), parseString),
		authors: retrieveAuthors(value),
		language: parseSingularOf(get("language"), parseString),
		attachments: parseArrayOf(get("attachments"), parseAttachment)
	});
};
const parseHub = (value) => {
	if (!isObject(value)) return;
	const get = createCaseInsensitiveGetter(value);
	return trimObject({
		type: parseSingularOf(get("type"), parseString),
		url: parseSingularOf(get("url"), parseString)
	});
};
const parseFeed = (value, options) => {
	if (!isObject(value)) return;
	const get = createCaseInsensitiveGetter(value);
	return trimObject({
		title: parseSingularOf(get("title"), parseString),
		home_page_url: parseSingularOf(get("home_page_url"), parseString),
		feed_url: parseSingularOf(get("feed_url"), parseString),
		description: parseSingularOf(get("description"), parseString),
		user_comment: parseSingularOf(get("user_comment"), parseString),
		next_url: parseSingularOf(get("next_url"), parseString),
		icon: parseSingularOf(get("icon"), parseString),
		favicon: parseSingularOf(get("favicon"), parseString),
		language: parseSingularOf(get("language"), parseString),
		expired: parseSingularOf(get("expired"), parseBoolean),
		hubs: parseArrayOf(get("hubs"), parseHub),
		authors: retrieveAuthors(value),
		items: parseArrayOf(get("items"), parseItem, options?.maxItems)
	});
};

//#endregion
export { createCaseInsensitiveGetter, parseFeed };
