// feedsmith/dist/namespaces/prism/parse/utils.js
import { isObject, parseArrayOf, parseDate, parseNumber, parseSingularOf, parseString, retrieveText, trimObject } from "../../../common/utils.js";

//#region src/namespaces/prism/parse/utils.ts
const retrieveFeed = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		publicationName: parseSingularOf(value["prism:publicationname"], (value$1) => parseString(retrieveText(value$1))),
		issn: parseSingularOf(value["prism:issn"], (value$1) => parseString(retrieveText(value$1))),
		eIssn: parseSingularOf(value["prism:eissn"], (value$1) => parseString(retrieveText(value$1))),
		isbns: parseArrayOf(value["prism:isbn"], (value$1) => parseString(retrieveText(value$1))),
		issueIdentifier: parseSingularOf(value["prism:issueidentifier"], (value$1) => parseString(retrieveText(value$1))),
		issueName: parseSingularOf(value["prism:issuename"], (value$1) => parseString(retrieveText(value$1))),
		issueTeaser: parseSingularOf(value["prism:issueteaser"], (value$1) => parseString(retrieveText(value$1))),
		issueType: parseSingularOf(value["prism:issuetype"], (value$1) => parseString(retrieveText(value$1))),
		volume: parseSingularOf(value["prism:volume"], (value$1) => parseString(retrieveText(value$1))),
		number: parseSingularOf(value["prism:number"], (value$1) => parseString(retrieveText(value$1))),
		edition: parseSingularOf(value["prism:edition"], (value$1) => parseString(retrieveText(value$1))),
		aggregateIssueNumber: parseSingularOf(value["prism:aggregateissuenumber"], (value$1) => parseNumber(retrieveText(value$1))),
		aggregationType: parseSingularOf(value["prism:aggregationtype"], (value$1) => parseString(retrieveText(value$1))),
		coverDate: parseSingularOf(value["prism:coverdate"], (value$1) => parseDate(retrieveText(value$1))),
		coverDisplayDate: parseSingularOf(value["prism:coverdisplaydate"], (value$1) => parseString(retrieveText(value$1))),
		publicationDates: parseArrayOf(value["prism:publicationdate"], (value$1) => parseDate(retrieveText(value$1))),
		publicationDisplayDates: parseArrayOf(value["prism:publicationdisplaydate"], (value$1) => parseString(retrieveText(value$1))),
		creationDate: parseSingularOf(value["prism:creationdate"], (value$1) => parseDate(retrieveText(value$1))),
		modificationDate: parseSingularOf(value["prism:modificationdate"], (value$1) => parseDate(retrieveText(value$1))),
		dateReceived: parseSingularOf(value["prism:datereceived"], (value$1) => parseDate(retrieveText(value$1))),
		onSaleDates: parseArrayOf(value["prism:onsaledate"], (value$1) => parseDate(retrieveText(value$1))),
		onSaleDays: parseArrayOf(value["prism:onsaleday"], (value$1) => parseString(retrieveText(value$1))),
		offSaleDates: parseArrayOf(value["prism:offsaledate"], (value$1) => parseDate(retrieveText(value$1))),
		killDate: parseSingularOf(value["prism:killdate"], (value$1) => parseDate(retrieveText(value$1))),
		copyrightYears: parseArrayOf(value["prism:copyrightyear"], (value$1) => parseString(retrieveText(value$1))),
		contentType: parseSingularOf(value["prism:contenttype"], (value$1) => parseString(retrieveText(value$1))),
		alternateTitles: parseArrayOf(value["prism:alternatetitle"], (value$1) => parseString(retrieveText(value$1))),
		subtitles: parseArrayOf(value["prism:subtitle"], (value$1) => parseString(retrieveText(value$1))),
		teasers: parseArrayOf(value["prism:teaser"], (value$1) => parseString(retrieveText(value$1))),
		keywords: parseArrayOf(value["prism:keyword"], (value$1) => parseString(retrieveText(value$1))),
		seriesTitle: parseSingularOf(value["prism:seriestitle"], (value$1) => parseString(retrieveText(value$1))),
		seriesNumber: parseSingularOf(value["prism:seriesnumber"], (value$1) => parseNumber(retrieveText(value$1))),
		bookEditions: parseArrayOf(value["prism:bookedition"], (value$1) => parseString(retrieveText(value$1))),
		nationalCatalogNumber: parseSingularOf(value["prism:nationalcatalognumber"], (value$1) => parseString(retrieveText(value$1))),
		productCodes: parseArrayOf(value["prism:productcode"], (value$1) => parseString(retrieveText(value$1))),
		uspsNumber: parseSingularOf(value["prism:uspsnumber"], (value$1) => parseString(retrieveText(value$1))),
		publishingFrequency: parseSingularOf(value["prism:publishingfrequency"], (value$1) => parseString(retrieveText(value$1))),
		channels: parseArrayOf(value["prism:channel"], (value$1) => parseString(retrieveText(value$1))),
		subchannel1: parseSingularOf(value["prism:subchannel1"], (value$1) => parseString(retrieveText(value$1))),
		subchannel2: parseSingularOf(value["prism:subchannel2"], (value$1) => parseString(retrieveText(value$1))),
		subchannel3: parseSingularOf(value["prism:subchannel3"], (value$1) => parseString(retrieveText(value$1))),
		subchannel4: parseSingularOf(value["prism:subchannel4"], (value$1) => parseString(retrieveText(value$1))),
		section: parseSingularOf(value["prism:section"], (value$1) => parseString(retrieveText(value$1))),
		subsection1: parseSingularOf(value["prism:subsection1"], (value$1) => parseString(retrieveText(value$1))),
		subsection2: parseSingularOf(value["prism:subsection2"], (value$1) => parseString(retrieveText(value$1))),
		subsection3: parseSingularOf(value["prism:subsection3"], (value$1) => parseString(retrieveText(value$1))),
		subsection4: parseSingularOf(value["prism:subsection4"], (value$1) => parseString(retrieveText(value$1))),
		corporateEntities: parseArrayOf(value["prism:corporateentity"], (value$1) => parseString(retrieveText(value$1))),
		distributor: parseSingularOf(value["prism:distributor"], (value$1) => parseString(retrieveText(value$1))),
		sellingAgencies: parseArrayOf(value["prism:sellingagency"], (value$1) => parseString(retrieveText(value$1))),
		organizations: parseArrayOf(value["prism:organization"], (value$1) => parseString(retrieveText(value$1))),
		persons: parseArrayOf(value["prism:person"], (value$1) => parseString(retrieveText(value$1))),
		platforms: parseArrayOf(value["prism:platform"], (value$1) => parseString(retrieveText(value$1))),
		originPlatforms: parseArrayOf(value["prism:originplatform"], (value$1) => parseString(retrieveText(value$1))),
		device: parseSingularOf(value["prism:device"], (value$1) => parseString(retrieveText(value$1))),
		complianceProfile: parseSingularOf(value["prism:complianceprofile"], (value$1) => parseString(retrieveText(value$1))),
		blogTitle: parseSingularOf(value["prism:blogtitle"], (value$1) => parseString(retrieveText(value$1))),
		blogURL: parseSingularOf(value["prism:blogurl"], (value$1) => parseString(retrieveText(value$1))),
		links: parseArrayOf(value["prism:link"], (value$1) => parseString(retrieveText(value$1))),
		urls: parseArrayOf(value["prism:url"], (value$1) => parseString(retrieveText(value$1))),
		byteCount: parseSingularOf(value["prism:bytecount"], (value$1) => parseNumber(retrieveText(value$1))),
		ratings: parseArrayOf(value["prism:rating"], (value$1) => parseString(retrieveText(value$1))),
		timePeriod: parseSingularOf(value["prism:timeperiod"], (value$1) => parseString(retrieveText(value$1))),
		versionIdentifier: parseSingularOf(value["prism:versionidentifier"], (value$1) => parseString(retrieveText(value$1))),
		tickers: parseArrayOf(value["prism:ticker"], (value$1) => parseString(retrieveText(value$1))),
		academicFields: parseArrayOf(value["prism:academicfield"], (value$1) => parseString(retrieveText(value$1))),
		events: parseArrayOf(value["prism:event"], (value$1) => parseString(retrieveText(value$1))),
		genres: parseArrayOf(value["prism:genre"], (value$1) => parseString(retrieveText(value$1))),
		industries: parseArrayOf(value["prism:industry"], (value$1) => parseString(retrieveText(value$1))),
		locations: parseArrayOf(value["prism:location"], (value$1) => parseString(retrieveText(value$1))),
		objects: parseArrayOf(value["prism:object"], (value$1) => parseString(retrieveText(value$1))),
		profession: parseSingularOf(value["prism:profession"], (value$1) => parseString(retrieveText(value$1))),
		sport: parseSingularOf(value["prism:sport"], (value$1) => parseString(retrieveText(value$1))),
		embargoDate: parseSingularOf(value["prism:embargodate"], (value$1) => parseDate(retrieveText(value$1))),
		copyright: parseSingularOf(value["prism:copyright"], (value$1) => parseString(retrieveText(value$1))),
		expirationDate: parseSingularOf(value["prism:expirationdate"], (value$1) => parseDate(retrieveText(value$1))),
		rightsAgent: parseSingularOf(value["prism:rightsagent"], (value$1) => parseString(retrieveText(value$1)))
	});
};
const retrieveItem = (value) => {
	if (!isObject(value)) return;
	return trimObject({
		publicationName: parseSingularOf(value["prism:publicationname"], (value$1) => parseString(retrieveText(value$1))),
		issn: parseSingularOf(value["prism:issn"], (value$1) => parseString(retrieveText(value$1))),
		eIssn: parseSingularOf(value["prism:eissn"], (value$1) => parseString(retrieveText(value$1))),
		doi: parseSingularOf(value["prism:doi"], (value$1) => parseString(retrieveText(value$1))),
		urls: parseArrayOf(value["prism:url"], (value$1) => parseString(retrieveText(value$1))),
		volume: parseSingularOf(value["prism:volume"], (value$1) => parseString(retrieveText(value$1))),
		number: parseSingularOf(value["prism:number"], (value$1) => parseString(retrieveText(value$1))),
		edition: parseSingularOf(value["prism:edition"], (value$1) => parseString(retrieveText(value$1))),
		section: parseSingularOf(value["prism:section"], (value$1) => parseString(retrieveText(value$1))),
		startingPage: parseSingularOf(value["prism:startingpage"], (value$1) => parseString(retrieveText(value$1))),
		endingPage: parseSingularOf(value["prism:endingpage"], (value$1) => parseString(retrieveText(value$1))),
		pageRange: parseSingularOf(value["prism:pagerange"], (value$1) => parseString(retrieveText(value$1))),
		pageCount: parseSingularOf(value["prism:pagecount"], (value$1) => parseNumber(retrieveText(value$1))),
		pageProgressionDirection: parseSingularOf(value["prism:pageprogressiondirection"], (value$1) => parseString(retrieveText(value$1))),
		samplePageRange: parseSingularOf(value["prism:samplepagerange"], (value$1) => parseString(retrieveText(value$1))),
		publicationDates: parseArrayOf(value["prism:publicationdate"], (value$1) => parseDate(retrieveText(value$1))),
		publicationDisplayDates: parseArrayOf(value["prism:publicationdisplaydate"], (value$1) => parseString(retrieveText(value$1))),
		creationDate: parseSingularOf(value["prism:creationdate"], (value$1) => parseDate(retrieveText(value$1))),
		modificationDate: parseSingularOf(value["prism:modificationdate"], (value$1) => parseDate(retrieveText(value$1))),
		dateReceived: parseSingularOf(value["prism:datereceived"], (value$1) => parseDate(retrieveText(value$1))),
		killDate: parseSingularOf(value["prism:killdate"], (value$1) => parseDate(retrieveText(value$1))),
		copyrightYears: parseArrayOf(value["prism:copyrightyear"], (value$1) => parseString(retrieveText(value$1))),
		contentType: parseSingularOf(value["prism:contenttype"], (value$1) => parseString(retrieveText(value$1))),
		genres: parseArrayOf(value["prism:genre"], (value$1) => parseString(retrieveText(value$1))),
		alternateTitles: parseArrayOf(value["prism:alternatetitle"], (value$1) => parseString(retrieveText(value$1))),
		subtitles: parseArrayOf(value["prism:subtitle"], (value$1) => parseString(retrieveText(value$1))),
		teasers: parseArrayOf(value["prism:teaser"], (value$1) => parseString(retrieveText(value$1))),
		keywords: parseArrayOf(value["prism:keyword"], (value$1) => parseString(retrieveText(value$1))),
		corporateEntities: parseArrayOf(value["prism:corporateentity"], (value$1) => parseString(retrieveText(value$1))),
		organizations: parseArrayOf(value["prism:organization"], (value$1) => parseString(retrieveText(value$1))),
		persons: parseArrayOf(value["prism:person"], (value$1) => parseString(retrieveText(value$1))),
		platforms: parseArrayOf(value["prism:platform"], (value$1) => parseString(retrieveText(value$1))),
		device: parseSingularOf(value["prism:device"], (value$1) => parseString(retrieveText(value$1))),
		academicFields: parseArrayOf(value["prism:academicfield"], (value$1) => parseString(retrieveText(value$1))),
		events: parseArrayOf(value["prism:event"], (value$1) => parseString(retrieveText(value$1))),
		industries: parseArrayOf(value["prism:industry"], (value$1) => parseString(retrieveText(value$1))),
		locations: parseArrayOf(value["prism:location"], (value$1) => parseString(retrieveText(value$1))),
		objects: parseArrayOf(value["prism:object"], (value$1) => parseString(retrieveText(value$1))),
		profession: parseSingularOf(value["prism:profession"], (value$1) => parseString(retrieveText(value$1))),
		sport: parseSingularOf(value["prism:sport"], (value$1) => parseString(retrieveText(value$1))),
		hasAlternatives: parseArrayOf(value["prism:hasalternative"], (value$1) => parseString(retrieveText(value$1))),
		hasCorrections: parseArrayOf(value["prism:hascorrection"], (value$1) => parseString(retrieveText(value$1))),
		hasTranslations: parseArrayOf(value["prism:hastranslation"], (value$1) => parseString(retrieveText(value$1))),
		isAlternativeOf: parseArrayOf(value["prism:isalternativeof"], (value$1) => parseString(retrieveText(value$1))),
		isCorrectionOf: parseArrayOf(value["prism:iscorrectionof"], (value$1) => parseString(retrieveText(value$1))),
		isTranslationOf: parseSingularOf(value["prism:istranslationof"], (value$1) => parseString(retrieveText(value$1))),
		supplementTitles: parseArrayOf(value["prism:supplementtitle"], (value$1) => parseString(retrieveText(value$1))),
		supplementDisplayID: parseSingularOf(value["prism:supplementdisplayid"], (value$1) => parseString(retrieveText(value$1))),
		supplementStartingPage: parseSingularOf(value["prism:supplementstartingpage"], (value$1) => parseString(retrieveText(value$1))),
		links: parseArrayOf(value["prism:link"], (value$1) => parseString(retrieveText(value$1))),
		wordCount: parseSingularOf(value["prism:wordcount"], (value$1) => parseNumber(retrieveText(value$1))),
		byteCount: parseSingularOf(value["prism:bytecount"], (value$1) => parseNumber(retrieveText(value$1))),
		ratings: parseArrayOf(value["prism:rating"], (value$1) => parseString(retrieveText(value$1))),
		timePeriod: parseSingularOf(value["prism:timeperiod"], (value$1) => parseString(retrieveText(value$1))),
		versionIdentifier: parseSingularOf(value["prism:versionidentifier"], (value$1) => parseString(retrieveText(value$1))),
		tickers: parseArrayOf(value["prism:ticker"], (value$1) => parseString(retrieveText(value$1))),
		embargoDate: parseSingularOf(value["prism:embargodate"], (value$1) => parseDate(retrieveText(value$1))),
		copyright: parseSingularOf(value["prism:copyright"], (value$1) => parseString(retrieveText(value$1))),
		expirationDate: parseSingularOf(value["prism:expirationdate"], (value$1) => parseDate(retrieveText(value$1))),
		rightsAgent: parseSingularOf(value["prism:rightsagent"], (value$1) => parseString(retrieveText(value$1)))
	});
};

//#endregion
export { retrieveFeed, retrieveItem };
